<!DOCTYPE html>
<html lang="en"><head>
<script src="presentationPartIV_files/libs/clipboard/clipboard.min.js"></script>
<script src="presentationPartIV_files/libs/quarto-html/tabby.min.js"></script>
<script src="presentationPartIV_files/libs/quarto-html/popper.min.js"></script>
<script src="presentationPartIV_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="presentationPartIV_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="presentationPartIV_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="presentationPartIV_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="presentationPartIV_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.47">
<meta name="author" content="Hemant Ishwaran and Min Lu   University of Miami">
<title>Tree-Based Machine Learning Methods for Prediction, Variable Selection and Causal Inference</title>
<meta name="apple-mobile-web-app-capable" content="yes">
<meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
<link rel="stylesheet" href="presentationPartIV_files/libs/revealjs/dist/reset.css">
<link rel="stylesheet" href="presentationPartIV_files/libs/revealjs/dist/reveal.css">
<style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
      }
    pre.numberSource { margin-left: 3em;  padding-left: 4px; }
    div.sourceCode
      { color: #cbccc6; background-color: #1f2430; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #cbccc6; } /* Normal */
    code span.al { color: #ff3333; background-color: #332430; font-weight: bold; } /* Alert */
    code span.an { color: #ffe6b3; } /* Annotation */
    code span.at { color: #73d0ff; } /* Attribute */
    code span.bn { color: #ffcc66; } /* BaseN */
    code span.bu { color: #95e6cb; } /* BuiltIn */
    code span.cf { color: #ffa759; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #95e6cb; } /* Char */
    code span.cn { color: #d4bfff; } /* Constant */
    code span.co { color: #5c6773; font-style: italic; } /* Comment */
    code span.cv { color: #d4bfff; } /* CommentVar */
    code span.do { color: #5c6773; } /* Documentation */
    code span.dt { color: #ffa759; } /* DataType */
    code span.dv { color: #ffcc66; } /* DecVal */
    code span.er { color: #ff3333; text-decoration: underline; } /* Error */
    code span.ex { color: #73d0ff; font-weight: bold; } /* Extension */
    code span.fl { color: #ffcc66; } /* Float */
    code span.fu { color: #ffd580; } /* Function */
    code span.im { color: #bae67e; } /* Import */
    code span.in { color: #ffcc66; } /* Information */
    code span.kw { color: #ffa759; font-weight: bold; } /* Keyword */
    code span.op { color: #f29e74; } /* Operator */
    code span.ot { color: #5ccfe6; } /* Other */
    code span.pp { color: #f28779; } /* Preprocessor */
    code span.re { color: #73d0ff; background-color: #2a4254; } /* RegionMarker */
    code span.sc { color: #95e6cb; } /* SpecialChar */
    code span.ss { color: #95e6cb; } /* SpecialString */
    code span.st { color: #bae67e; } /* String */
    code span.va { color: #5ccfe6; } /* Variable */
    code span.vs { color: #bae67e; } /* VerbatimString */
    code span.wa { color: #f28779; } /* Warning */
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
<link rel="stylesheet" href="presentationPartIV_files/libs/revealjs/dist/theme/quarto.css">
<link rel="stylesheet" href="styles.css">
<link href="presentationPartIV_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
<link href="presentationPartIV_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
<link href="presentationPartIV_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
<link href="presentationPartIV_files/libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
<link href="presentationPartIV_files/libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
<link href="presentationPartIV_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
<style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    padding-bottom: 0.5rem;
    margin-bottom: 0;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
<style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" data-background-image="images/forest_mathextra_book.jpg" data-background-opacity="0.95" class="quarto-title-block center"><h1 class="title">Tree-Based Machine Learning Methods for Prediction, Variable Selection and Causal Inference</h1>
  <p class="subtitle">Part IV</p>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Hemant Ishwaran and Min Lu <br> University of Miami 
</div>
</div>
</div>

</section><section id="advanced-examples" class="slide level2"><h2>Advanced examples</h2>
<ol type="1">
<li>Class imbalanced data</li>
<li>Competing Risks</li>
<li>Multivariate Forests</li>
</ol></section><section id="imbalanced-classification" class="slide level2 smaller"><h2>Imbalanced classification</h2>
<p><br></p>
<table class="table-striped border small table-sm caption-top">
<colgroup>
<col style="width: 52%">
<col style="width: 47%">
</colgroup>
<thead><tr class="header">
<th>Family</th>
<th>Example Grow Call with Formula Specification</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><span style="color:#00589b;">Regression <br> Quantile Regression</span></td>
<td>
<code>rfsrc(Ozone~., data = airquality)</code> <br><code>quantreg(mpg~., data = mtcars)</code>
</td>
</tr>
<tr class="even">
<td>
<span style="color:#00589b;">Classification</span> <br><span class="fragment highlight-blue" data-fragment-index="1">Imbalanced Two-Class</span>
</td>
<td>rfsrc(Species~., data = iris) <br><span class="fragment highlight-blue" data-fragment-index="1">imbalanced(status~., data = breast)</span>
</td>
</tr>
<tr class="odd">
<td><span style="color:#00589b;">Survival</span></td>
<td><code>rfsrc(Surv(time, status)~., data = veteran)</code></td>
</tr>
<tr class="even">
<td>Competing Risk</td>
<td><code>rfsrc(Surv(time, status)~., data = wihs)</code></td>
</tr>
<tr class="odd">
<td>Multivariate Regression <br> Multivariate Mixed Regression <br> Multivariate Quantile Regression <br> Multivariate Mixed Quantile Regression</td>
<td>
<code>rfsrc(Multivar(mpg, cyl)~., data = mtcars)</code> <br><code>rfsrc(cbind(Species,Sepal.Length)~.,data=iris)</code> <br><code>quantreg(cbind(mpg, cyl)~., data = mtcars)</code> <br><code>quantreg(cbind(Species,Sepal.Length)~.,data=iris)</code>
</td>
</tr>
<tr class="even">
<td>Unsupervised <br> sidClustering <br> Breiman (Shi-Horvath)</td>
<td>
<code>rfsrc(data = mtcars)</code> <br><code>sidClustering(data = mtcars)</code> <br><code>sidClustering(data = mtcars, method = "sh")</code>
</td>
</tr>
</tbody>
</table></section><section id="class-imbalanced-data" class="slide level2 smaller"><h2>Class imbalanced data</h2>
<p>In many real world applications, a common problem is the occurrence of imbalanced data where one class outcome (majority class, 0) is observed far more frequently than the other (minority class, 1)</p>
<div>
<ol type="1">
<li class="fragment">Business analytics: customer churn data</li>
<li class="fragment">Software and text recognition: spam filtering, detecting coding errors</li>
<li class="fragment">Fault detection and quality control in industrial systems</li>
<li class="fragment">Biomedical data when prevelance of target group is relatively small or event measured over short interval</li>
</ol>
</div>
<div class="footer">
<p>Learn more <span class="citation" data-cites="HemantRFQv">[<a href="#/references" role="doc-biblioref" onclick="">1</a>]</span>: <a href="https://www.randomforestsrc.org/articles/imbalance.html">https://www.randomforestsrc.org/articles/imbalance.html</a></p>
</div>
</section><section id="example-from-causal-analysis" class="slide level2 smaller"><h2>Example from causal analysis</h2>
<p>An investigator studying the value of adjuvant treatment versus surgery in esophageal patients needed to apply causal techniques for their analysis. This required predicting the treatment group using patient features (treatment overlap analysis). There are <span class="math inline">\(N_0=6649\)</span> surgery patients and <span class="math inline">\(N_1=988\)</span> adjuvant treatment patients, giving an imbalanced ratio of <span class="math inline">\(6649/988 \approx 6.8\)</span>.</p>
<p>Running a RF classifier on the data, they obtained the following confusion matrix (from the OOB ensemble): <br></p>
<table class="table-striped border small table-sm caption-top">
<colgroup>
<col style="width: 24%">
<col style="width: 28%">
<col style="width: 24%">
<col style="width: 22%">
</colgroup>
<thead><tr class="header">
<th></th>
<th></th>
<th>predicted</th>
<th></th>
</tr></thead>
<tbody>
<tr class="odd">
<td></td>
<td>surgery</td>
<td>adjuvant</td>
<td>error</td>
</tr>
<tr class="even">
<td>surgery</td>
<td>6549</td>
<td>100</td>
<td>.015</td>
</tr>
<tr class="odd">
<td>adjuvant</td>
<td>694</td>
<td>294</td>
<td>.702</td>
</tr>
</tbody>
</table>
<center>
Overall error rate: 12.2%
</center>
</section><section id="example-from-causal-analysis-1" class="slide level2 smaller"><h2>Example from causal analysis</h2>
<p>Running a RF classifier on the data, they obtained the following confusion matrix (from the OOB ensemble): <br></p>
<table class="table-striped border small table-sm caption-top">
<colgroup>
<col style="width: 24%">
<col style="width: 28%">
<col style="width: 24%">
<col style="width: 22%">
</colgroup>
<thead><tr class="header">
<th></th>
<th></th>
<th>predicted</th>
<th></th>
</tr></thead>
<tbody>
<tr class="odd">
<td></td>
<td>surgery</td>
<td>adjuvant</td>
<td>error</td>
</tr>
<tr class="even">
<td>surgery</td>
<td>6549</td>
<td>100</td>
<td>.015</td>
</tr>
<tr class="odd">
<td>adjuvant</td>
<td>694</td>
<td>294</td>
<td>.702</td>
</tr>
</tbody>
</table>
<center>
Overall error rate: 12.2%
</center>
<div class="fragment">
<p>The TNR (true postive rate) and TPR (false positive rate) are: <span class="math display">\[
\text{TNR} = \frac{6549}{6549+100}=98.5\%,\hskip10pt
\text{TPR} = \frac{294}{694+294}=29.8\%,\hskip10pt
\]</span> The geometric mean (gmean) is <span class="math display">\[
\text{gmean}=\sqrt{.985\times .298} = 54.2\%
\]</span> This is very poor, random guessing gives 0.5</p>
</div>
</section><section id="metrics-for-evaluating-class-imbalanced-performance" class="slide level2 smaller"><h2>Metrics for evaluating class imbalanced performance</h2>
<p>Avoid using AUC, misclassification error and other standard metrics and instead use:</p>
<ol type="1">
<li>gmean</li>
<li>PR-AUC (precision recall, AUC)</li>
</ol>
<center>
<img data-src="pic/rfq_adjuvantEsophagus.png" style="width:50.0%"><br><p style="font-family:arcade classic;font-size:20px">
ROC curve and PR-AUC curve for RF classifier from esophageal study of adjuvant treatment versus surgery
</p>
</center>
<div class="footer">

</div>
</section><section id="why-does-the-classifier-have-such-poor-performance" class="slide level2"><h2>Why does the classifier have such poor performance?</h2>
<p>The problem here (and for other ML methods) is the equal misclassification costs used by the Bayes decision rule <span class="math display">\[
\delta(\mathbf{x})=1 \hskip5pt\text{if}\hskip5pt P\{Y = 1 \mid \mathbf{X} = \mathbf{x}\} \ge \frac{1}{2}
\]</span></p>
<p>In imbalanced data, the probability of an event is very small!! Therefore, the classifier tends to classify all cases as class labels 0</p>
</section><section id="rfq-random-forest-quantile-classifier" class="slide level2"><h2>RFQ (random forest quantile) classifier</h2>
<p>Use a different threshold. In place of 0.5 (the median) we use <span class="math inline">\(\pi=P\{Y=1\}\)</span> <span class="math display">\[
\delta(\mathbf{x})=1
\hskip5pt\text{if}\hskip5pt
P\{Y = 1 \mid \mathbf{X} = \mathbf{x}\}
\ge
\pi
\approx \frac{N_1}{N_0+N_1}
\]</span></p>
<p>This has a dual optimality property (cost-weighted Bayes optimal and maximizes TNR+TPR)</p>
</section><section id="general-call-to-imbalanced" class="slide level2"><h2>General call to imbalanced</h2>
<div class="fragment fade-left">
<p><img data-src="pic/imbalanced.png" style="width:100.0%"></p>
</div>
</section><section id="classification-example-glioma" class="slide level2 smaller"><h2>Classification example: Glioma</h2>
<h3 id="imbalanced-class">Imbalanced class</h3>
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Tip</strong></p>
</div>
<div class="callout-content">
<p><code>imbalanced()</code> is useful if you have two imbalanced class labels</p>
</div>
</div>
</div>
<div class="sourceCode" id="cb1" data-code-line-numbers="4|6-9|10-15"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href=""></a><span class="fu">data</span>(glioma, <span class="at">package =</span> <span class="st">"varPro"</span>)</span>
<span id="cb1-2"><a href=""></a><span class="fu">table</span>(glioma<span class="sc">$</span>y)</span>
<span id="cb1-3"><a href=""></a><span class="sc">&gt;</span>     Classic<span class="sc">-</span>like            Codel      G<span class="sc">-</span>CIMP<span class="sc">-</span>high       G<span class="sc">-</span>CIMP<span class="sc">-</span>low         LGm6<span class="sc">-</span>GBM Mesenchymal<span class="sc">-</span>like          PA<span class="sc">-</span>like </span>
<span id="cb1-4"><a href=""></a><span class="sc">&gt;</span>              <span class="dv">148</span>              <span class="dv">174</span>              <span class="dv">249</span>               <span class="dv">25</span>               <span class="dv">41</span>              <span class="dv">215</span>               <span class="dv">26</span> </span>
<span id="cb1-5"><a href=""></a></span>
<span id="cb1-6"><a href=""></a><span class="do">## make a super majority class</span></span>
<span id="cb1-7"><a href=""></a>class.combine <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Classic-like"</span>, <span class="st">"Codel"</span>, <span class="st">"G-CIMP-high"</span>, <span class="st">"Mesenchymal-like"</span>)</span>
<span id="cb1-8"><a href=""></a>ynew <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="dv">1</span> <span class="sc">*</span> <span class="sc">!</span><span class="fu">is.element</span>(glioma<span class="sc">$</span>y, class.combine))</span>
<span id="cb1-9"><a href=""></a></span>
<span id="cb1-10"><a href=""></a><span class="do">## replace y with the new super class</span></span>
<span id="cb1-11"><a href=""></a>glioma2 <span class="ot">&lt;-</span> glioma</span>
<span id="cb1-12"><a href=""></a>glioma2<span class="sc">$</span>y <span class="ot">&lt;-</span> ynew</span>
<span id="cb1-13"><a href=""></a><span class="fu">table</span>(glioma2<span class="sc">$</span>y)</span>
<span id="cb1-14"><a href=""></a><span class="sc">&gt;</span>    <span class="dv">0</span>   <span class="dv">1</span> </span>
<span id="cb1-15"><a href=""></a><span class="sc">&gt;</span>  <span class="dv">786</span>  <span class="dv">94</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section><section id="imbalanced-classification-glioma" class="slide level2 smaller"><h2>Imbalanced classification: Glioma</h2>
<h4 id="standard-rf-classifier">standard RF classifier</h4>
<div class="sourceCode" id="cb2" data-code-line-numbers="1-3|13-14|17|18-24|26-33"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href=""></a><span class="do">## standard RF classifier</span></span>
<span id="cb2-2"><a href=""></a>o1 <span class="ot">&lt;-</span> <span class="fu">rfsrc</span>(y<span class="sc">~</span>.,glioma2)</span>
<span id="cb2-3"><a href=""></a><span class="fu">print</span>(o1)</span>
<span id="cb2-4"><a href=""></a>                         Sample size<span class="sc">:</span> <span class="dv">880</span></span>
<span id="cb2-5"><a href=""></a>           Frequency of class labels<span class="sc">:</span> <span class="dv">786</span>, <span class="dv">94</span></span>
<span id="cb2-6"><a href=""></a>                     Number of trees<span class="sc">:</span> <span class="dv">500</span></span>
<span id="cb2-7"><a href=""></a>           Forest terminal node size<span class="sc">:</span> <span class="dv">1</span></span>
<span id="cb2-8"><a href=""></a>       Average no. of terminal nodes<span class="sc">:</span> <span class="fl">29.726</span></span>
<span id="cb2-9"><a href=""></a>No. of variables tried at each split<span class="sc">:</span> <span class="dv">36</span></span>
<span id="cb2-10"><a href=""></a>              Total no. of variables<span class="sc">:</span> <span class="dv">1241</span></span>
<span id="cb2-11"><a href=""></a>       Resampling used to grow trees<span class="sc">:</span> swor</span>
<span id="cb2-12"><a href=""></a>    Resample size used to grow trees<span class="sc">:</span> <span class="dv">556</span></span>
<span id="cb2-13"><a href=""></a>                            Analysis<span class="sc">:</span> RF<span class="sc">-</span>C</span>
<span id="cb2-14"><a href=""></a>                              Family<span class="sc">:</span> class</span>
<span id="cb2-15"><a href=""></a>                      Splitting rule<span class="sc">:</span> gini <span class="sc">*</span>random<span class="sc">*</span></span>
<span id="cb2-16"><a href=""></a>       Number of random split points<span class="sc">:</span> <span class="dv">10</span></span>
<span id="cb2-17"><a href=""></a>                    Imbalanced ratio<span class="sc">:</span> <span class="fl">8.3617</span></span>
<span id="cb2-18"><a href=""></a>                   (OOB) Brier score<span class="sc">:</span> <span class="fl">0.03639723</span></span>
<span id="cb2-19"><a href=""></a>        (OOB) Normalized Brier score<span class="sc">:</span> <span class="fl">0.1455889</span></span>
<span id="cb2-20"><a href=""></a>                           (OOB) AUC<span class="sc">:</span> <span class="fl">0.98621488</span></span>
<span id="cb2-21"><a href=""></a>                      (OOB) Log<span class="sc">-</span>loss<span class="sc">:</span> <span class="fl">0.1298745</span></span>
<span id="cb2-22"><a href=""></a>                        (OOB) PR<span class="sc">-</span>AUC<span class="sc">:</span> <span class="fl">0.9170852</span></span>
<span id="cb2-23"><a href=""></a>                        (OOB) G<span class="sc">-</span>mean<span class="sc">:</span> <span class="fl">0.72932496</span></span>
<span id="cb2-24"><a href=""></a>   (OOB) Requested performance error<span class="sc">:</span> <span class="fl">0.05</span>, <span class="dv">0</span>, <span class="fl">0.46808511</span></span>
<span id="cb2-25"><a href=""></a></span>
<span id="cb2-26"><a href=""></a>Confusion matrix<span class="sc">:</span></span>
<span id="cb2-27"><a href=""></a></span>
<span id="cb2-28"><a href=""></a>          predicted</span>
<span id="cb2-29"><a href=""></a>  observed   <span class="dv">0</span>  <span class="dv">1</span> class.error</span>
<span id="cb2-30"><a href=""></a>         <span class="dv">0</span> <span class="dv">786</span>  <span class="dv">0</span>      <span class="fl">0.0000</span></span>
<span id="cb2-31"><a href=""></a>         <span class="dv">1</span>  <span class="dv">44</span> <span class="dv">50</span>      <span class="fl">0.4681</span></span>
<span id="cb2-32"><a href=""></a></span>
<span id="cb2-33"><a href=""></a>      (OOB) Misclassification rate<span class="sc">:</span> <span class="fl">0.05</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section><section id="imbalanced-classification-glioma-1" class="slide level2 smaller"><h2>Imbalanced classification: Glioma</h2>
<h4 id="rfq-classifier">RFQ classifier</h4>
<div class="sourceCode" id="cb3" data-code-line-numbers="1-3|13-14|17|18-24|26-33"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href=""></a><span class="do">## RFQ classifier</span></span>
<span id="cb3-2"><a href=""></a>o2 <span class="ot">&lt;-</span> <span class="fu">imbalanced</span>(y<span class="sc">~</span>.,glioma2)</span>
<span id="cb3-3"><a href=""></a><span class="fu">print</span>(o2)</span>
<span id="cb3-4"><a href=""></a>                         Sample size<span class="sc">:</span> <span class="dv">880</span></span>
<span id="cb3-5"><a href=""></a>           Frequency of class labels<span class="sc">:</span> <span class="dv">786</span>, <span class="dv">94</span></span>
<span id="cb3-6"><a href=""></a>                     Number of trees<span class="sc">:</span> <span class="dv">3000</span></span>
<span id="cb3-7"><a href=""></a>           Forest terminal node size<span class="sc">:</span> <span class="dv">1</span></span>
<span id="cb3-8"><a href=""></a>       Average no. of terminal nodes<span class="sc">:</span> <span class="fl">26.557</span></span>
<span id="cb3-9"><a href=""></a>No. of variables tried at each split<span class="sc">:</span> <span class="dv">36</span></span>
<span id="cb3-10"><a href=""></a>              Total no. of variables<span class="sc">:</span> <span class="dv">1241</span></span>
<span id="cb3-11"><a href=""></a>       Resampling used to grow trees<span class="sc">:</span> swor</span>
<span id="cb3-12"><a href=""></a>    Resample size used to grow trees<span class="sc">:</span> <span class="dv">556</span></span>
<span id="cb3-13"><a href=""></a>                            Analysis<span class="sc">:</span> RFQ</span>
<span id="cb3-14"><a href=""></a>                              Family<span class="sc">:</span> class</span>
<span id="cb3-15"><a href=""></a>                      Splitting rule<span class="sc">:</span> auc <span class="sc">*</span>random<span class="sc">*</span></span>
<span id="cb3-16"><a href=""></a>       Number of random split points<span class="sc">:</span> <span class="dv">10</span></span>
<span id="cb3-17"><a href=""></a>                    Imbalanced ratio<span class="sc">:</span> <span class="fl">8.3617</span></span>
<span id="cb3-18"><a href=""></a>                   (OOB) Brier score<span class="sc">:</span> <span class="fl">0.03362459</span></span>
<span id="cb3-19"><a href=""></a>        (OOB) Normalized Brier score<span class="sc">:</span> <span class="fl">0.13449834</span></span>
<span id="cb3-20"><a href=""></a>                           (OOB) AUC<span class="sc">:</span> <span class="fl">0.99057983</span></span>
<span id="cb3-21"><a href=""></a>                      (OOB) Log<span class="sc">-</span>loss<span class="sc">:</span> <span class="fl">0.12037099</span></span>
<span id="cb3-22"><a href=""></a>                        (OOB) PR<span class="sc">-</span>AUC<span class="sc">:</span> <span class="fl">0.94065642</span></span>
<span id="cb3-23"><a href=""></a>                        (OOB) G<span class="sc">-</span>mean<span class="sc">:</span> <span class="fl">0.93830013</span></span>
<span id="cb3-24"><a href=""></a>   (OOB) Requested performance error<span class="sc">:</span> <span class="fl">0.06169987</span></span>
<span id="cb3-25"><a href=""></a></span>
<span id="cb3-26"><a href=""></a>Confusion matrix<span class="sc">:</span></span>
<span id="cb3-27"><a href=""></a></span>
<span id="cb3-28"><a href=""></a>          predicted</span>
<span id="cb3-29"><a href=""></a>  observed   <span class="dv">0</span>  <span class="dv">1</span> class.error</span>
<span id="cb3-30"><a href=""></a>         <span class="dv">0</span> <span class="dv">692</span> <span class="dv">94</span>      <span class="fl">0.1196</span></span>
<span id="cb3-31"><a href=""></a>         <span class="dv">1</span>   <span class="dv">0</span> <span class="dv">94</span>      <span class="fl">0.0000</span></span>
<span id="cb3-32"><a href=""></a></span>
<span id="cb3-33"><a href=""></a>      (OOB) Misclassification rate<span class="sc">:</span> <span class="fl">0.1068182</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section><section id="imbalanced-classification-glioma-2" class="slide level2"><h2>Imbalanced classification: Glioma</h2>
<h4 id="balanced-rf-brf">Balanced RF (BRF)</h4>
<ol type="1">
<li>BRF under-samples the majority class to balance the frequencies</li>
<li>This is an example of a data-level method</li>
<li>Surprisingly, BRF is actually an example of a quantile classifier and is <em>theoretically</em> equivalent to RFQ</li>
<li>However, in finite samples it is less efficient</li>
</ol></section><section id="imbalanced-classification-glioma-3" class="slide level2 smaller"><h2>Imbalanced classification: Glioma</h2>
<div class="sourceCode" id="cb4" data-code-line-numbers="1-3|13-14|17|18-24|26-33"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href=""></a><span class="do">## BRF classifier</span></span>
<span id="cb4-2"><a href=""></a>o3 <span class="ot">&lt;-</span> <span class="fu">imbalanced</span>(y<span class="sc">~</span>.,glioma2, <span class="at">method =</span> <span class="st">"brf"</span>)</span>
<span id="cb4-3"><a href=""></a><span class="fu">print</span>(o3)</span>
<span id="cb4-4"><a href=""></a>                         Sample size<span class="sc">:</span> <span class="dv">880</span></span>
<span id="cb4-5"><a href=""></a>           Frequency of class labels<span class="sc">:</span> <span class="dv">786</span>, <span class="dv">94</span></span>
<span id="cb4-6"><a href=""></a>                     Number of trees<span class="sc">:</span> <span class="dv">3000</span></span>
<span id="cb4-7"><a href=""></a>           Forest terminal node size<span class="sc">:</span> <span class="dv">1</span></span>
<span id="cb4-8"><a href=""></a>       Average no. of terminal nodes<span class="sc">:</span> <span class="fl">15.1263</span></span>
<span id="cb4-9"><a href=""></a>No. of variables tried at each split<span class="sc">:</span> <span class="dv">36</span></span>
<span id="cb4-10"><a href=""></a>              Total no. of variables<span class="sc">:</span> <span class="dv">1241</span></span>
<span id="cb4-11"><a href=""></a>       Resampling used to grow trees<span class="sc">:</span> swr</span>
<span id="cb4-12"><a href=""></a>    Resample size used to grow trees<span class="sc">:</span> <span class="dv">188</span></span>
<span id="cb4-13"><a href=""></a>                            Analysis<span class="sc">:</span> RF<span class="sc">-</span>C</span>
<span id="cb4-14"><a href=""></a>                              Family<span class="sc">:</span> class</span>
<span id="cb4-15"><a href=""></a>                      Splitting rule<span class="sc">:</span> auc <span class="sc">*</span>random<span class="sc">*</span></span>
<span id="cb4-16"><a href=""></a>       Number of random split points<span class="sc">:</span> <span class="dv">10</span></span>
<span id="cb4-17"><a href=""></a>                    Imbalanced ratio<span class="sc">:</span> <span class="fl">8.3617</span></span>
<span id="cb4-18"><a href=""></a>                   (OOB) Brier score<span class="sc">:</span> <span class="fl">0.04844594</span></span>
<span id="cb4-19"><a href=""></a>        (OOB) Normalized Brier score<span class="sc">:</span> <span class="fl">0.19378376</span></span>
<span id="cb4-20"><a href=""></a>                           (OOB) AUC<span class="sc">:</span> <span class="fl">0.98658708</span></span>
<span id="cb4-21"><a href=""></a>                      (OOB) Log<span class="sc">-</span>loss<span class="sc">:</span> <span class="fl">0.19773628</span></span>
<span id="cb4-22"><a href=""></a>                        (OOB) PR<span class="sc">-</span>AUC<span class="sc">:</span> <span class="fl">0.92307335</span></span>
<span id="cb4-23"><a href=""></a>                        (OOB) G<span class="sc">-</span>mean<span class="sc">:</span> <span class="fl">0.91219607</span></span>
<span id="cb4-24"><a href=""></a>   (OOB) Requested performance error<span class="sc">:</span> <span class="fl">0.08780393</span></span>
<span id="cb4-25"><a href=""></a></span>
<span id="cb4-26"><a href=""></a>Confusion matrix<span class="sc">:</span></span>
<span id="cb4-27"><a href=""></a></span>
<span id="cb4-28"><a href=""></a>          predicted</span>
<span id="cb4-29"><a href=""></a>  observed   <span class="dv">0</span>  <span class="dv">1</span> class.error</span>
<span id="cb4-30"><a href=""></a>         <span class="dv">0</span> <span class="dv">759</span> <span class="dv">27</span>      <span class="fl">0.0344</span></span>
<span id="cb4-31"><a href=""></a>         <span class="dv">1</span>  <span class="dv">13</span> <span class="dv">81</span>      <span class="fl">0.1383</span></span>
<span id="cb4-32"><a href=""></a></span>
<span id="cb4-33"><a href=""></a>      (OOB) Misclassification rate<span class="sc">:</span> <span class="fl">0.04545455</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section><section id="imbalanced-classification-glioma-4" class="slide level2 smaller"><h2>Imbalanced classification: Glioma</h2>
<div class="sourceCode" id="cb5" data-code-line-numbers="1-5|15-16|19|20-26|28-35"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href=""></a><span class="do">## BRF - brute force</span></span>
<span id="cb5-2"><a href=""></a>o4 <span class="ot">&lt;-</span> <span class="fu">rfsrc</span>(y<span class="sc">~</span>.,glioma2,</span>
<span id="cb5-3"><a href=""></a>            <span class="at">case.wt =</span> randomForestSRC<span class="sc">:::</span><span class="fu">make.wt</span>(glioma2<span class="sc">$</span>y),</span>
<span id="cb5-4"><a href=""></a>            <span class="at">sampsize =</span> randomForestSRC<span class="sc">:::</span><span class="fu">make.size</span>(glioma2<span class="sc">$</span>y))</span>
<span id="cb5-5"><a href=""></a><span class="fu">print</span>(o4)</span>
<span id="cb5-6"><a href=""></a>                         Sample size<span class="sc">:</span> <span class="dv">880</span></span>
<span id="cb5-7"><a href=""></a>           Frequency of class labels<span class="sc">:</span> <span class="dv">786</span>, <span class="dv">94</span></span>
<span id="cb5-8"><a href=""></a>                     Number of trees<span class="sc">:</span> <span class="dv">500</span></span>
<span id="cb5-9"><a href=""></a>           Forest terminal node size<span class="sc">:</span> <span class="dv">1</span></span>
<span id="cb5-10"><a href=""></a>       Average no. of terminal nodes<span class="sc">:</span> <span class="fl">19.234</span></span>
<span id="cb5-11"><a href=""></a>No. of variables tried at each split<span class="sc">:</span> <span class="dv">36</span></span>
<span id="cb5-12"><a href=""></a>              Total no. of variables<span class="sc">:</span> <span class="dv">1241</span></span>
<span id="cb5-13"><a href=""></a>       Resampling used to grow trees<span class="sc">:</span> swor</span>
<span id="cb5-14"><a href=""></a>    Resample size used to grow trees<span class="sc">:</span> <span class="dv">188</span></span>
<span id="cb5-15"><a href=""></a>                            Analysis<span class="sc">:</span> RF<span class="sc">-</span>C</span>
<span id="cb5-16"><a href=""></a>                              Family<span class="sc">:</span> class</span>
<span id="cb5-17"><a href=""></a>                      Splitting rule<span class="sc">:</span> gini <span class="sc">*</span>random<span class="sc">*</span></span>
<span id="cb5-18"><a href=""></a>       Number of random split points<span class="sc">:</span> <span class="dv">10</span></span>
<span id="cb5-19"><a href=""></a>                    Imbalanced ratio<span class="sc">:</span> <span class="fl">8.3617</span></span>
<span id="cb5-20"><a href=""></a>                   (OOB) Brier score<span class="sc">:</span> <span class="fl">0.05301977</span></span>
<span id="cb5-21"><a href=""></a>        (OOB) Normalized Brier score<span class="sc">:</span> <span class="fl">0.21207909</span></span>
<span id="cb5-22"><a href=""></a>                           (OOB) AUC<span class="sc">:</span> <span class="fl">0.98412376</span></span>
<span id="cb5-23"><a href=""></a>                      (OOB) Log<span class="sc">-</span>loss<span class="sc">:</span> <span class="fl">0.20939848</span></span>
<span id="cb5-24"><a href=""></a>                        (OOB) PR<span class="sc">-</span>AUC<span class="sc">:</span> <span class="fl">0.90805735</span></span>
<span id="cb5-25"><a href=""></a>                        (OOB) G<span class="sc">-</span>mean<span class="sc">:</span> <span class="fl">0.90192259</span></span>
<span id="cb5-26"><a href=""></a>   (OOB) Requested performance error<span class="sc">:</span> <span class="fl">0.06477273</span>, <span class="fl">0.05597964</span>, <span class="fl">0.13829787</span></span>
<span id="cb5-27"><a href=""></a></span>
<span id="cb5-28"><a href=""></a>Confusion matrix<span class="sc">:</span></span>
<span id="cb5-29"><a href=""></a></span>
<span id="cb5-30"><a href=""></a>          predicted</span>
<span id="cb5-31"><a href=""></a>  observed   <span class="dv">0</span>  <span class="dv">1</span> class.error</span>
<span id="cb5-32"><a href=""></a>         <span class="dv">0</span> <span class="dv">742</span> <span class="dv">44</span>      <span class="fl">0.0560</span></span>
<span id="cb5-33"><a href=""></a>         <span class="dv">1</span>  <span class="dv">13</span> <span class="dv">81</span>      <span class="fl">0.1383</span></span>
<span id="cb5-34"><a href=""></a></span>
<span id="cb5-35"><a href=""></a>      (OOB) Misclassification rate<span class="sc">:</span> <span class="fl">0.06477273</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section><section id="imbalanced-classification-glioma-5" class="slide level2 smaller"><h2>Imbalanced classification: Glioma</h2>
<div class="sourceCode" id="cb6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href=""></a><span class="do">## gmean variable importance with ci</span></span>
<span id="cb6-2"><a href=""></a>o2 <span class="ot">&lt;-</span> <span class="fu">imbalanced</span>(y<span class="sc">~</span>.,glioma2, <span class="at">importance=</span><span class="st">"permute"</span>, <span class="at">block.size=</span><span class="dv">20</span>)</span>
<span id="cb6-3"><a href=""></a>oo2 <span class="ot">&lt;-</span> <span class="fu">subsample</span>(o2)</span>
<span id="cb6-4"><a href=""></a><span class="fu">plot.subsample</span>(oo2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="fragment fade-left">
<p><img data-src="pic/rfq_gmean_vs_varpro.png" style="width:82.0%"></p>
</div>
<div class="footer">

</div>
</section><section id="competing-risk" class="slide level2 smaller"><h2>Competing risk</h2>
<p><br></p>
<table class="table-striped border small table-sm caption-top">
<colgroup>
<col style="width: 52%">
<col style="width: 47%">
</colgroup>
<thead><tr class="header">
<th>Family</th>
<th>Example Grow Call with Formula Specification</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><span style="color:#00589b;">Regression <br> Quantile Regression</span></td>
<td>
<code>rfsrc(Ozone~., data = airquality)</code> <br><code>quantreg(mpg~., data = mtcars)</code>
</td>
</tr>
<tr class="even">
<td>
<span style="color:#00589b;">Classification</span> <br><span style="color:#00589b;">Imbalanced Two-Class</span>
</td>
<td>
<code>rfsrc(Species~., data = iris)</code> <br><code>imbalanced(status~., data = breast)</code>
</td>
</tr>
<tr class="odd">
<td><span style="color:#00589b;">Survival</span></td>
<td><code>rfsrc(Surv(time, status)~., data = veteran)</code></td>
</tr>
<tr class="even">
<td><span class="fragment highlight-blue" data-fragment-index="1">Competing Risk</span></td>
<td><span class="fragment highlight-blue" data-fragment-index="1">rfsrc(Surv(time, status)~., data = wihs)</span></td>
</tr>
<tr class="odd">
<td>Multivariate Regression <br> Multivariate Mixed Regression <br> Multivariate Quantile Regression <br> Multivariate Mixed Quantile Regression</td>
<td>
<code>rfsrc(Multivar(mpg, cyl)~., data = mtcars)</code> <br><code>rfsrc(cbind(Species,Sepal.Length)~.,data=iris)</code> <br><code>quantreg(cbind(mpg, cyl)~., data = mtcars)</code> <br><code>quantreg(cbind(Species,Sepal.Length)~.,data=iris)</code>
</td>
</tr>
<tr class="even">
<td>Unsupervised <br> sidClustering <br> Breiman (Shi-Horvath)</td>
<td>
<code>rfsrc(data = mtcars)</code> <br><code>sidClustering(data = mtcars)</code> <br><code>sidClustering(data = mtcars, method = "sh")</code>
</td>
</tr>
</tbody>
</table></section><section id="competing-risk-1" class="slide level2 smaller"><h2>Competing risk</h2>
<p>This is a specialized survival (time to event) analysis setting:</p>
<ol type="1">
<li>In competing risk the individual is subject to more than one event.</li>
<li>For example, a patient listed to receive a heart transplant can die while on the waitlist, or they can receive the heart. The patient is subject to two competing risks “death” and “transplant”</li>
<li>Right-censoring also occurs just like in standard survival analysis</li>
</ol>
<div class="footer">
<p>Learn more <span class="citation" data-cites="HemantCompeting">[<a href="#/references" role="doc-biblioref" onclick="">2</a>]</span>: <a href="https://www.randomforestsrc.org/articles/competing.html">https://www.randomforestsrc.org/articles/competing.html</a></p>
</div>
</section><section id="competing-risk-quantities-of-interest" class="slide level2 smaller"><h2>Competing risk quantities of interest</h2>
<p>There are two primary quantities of interest:</p>
<h4 id="the-cumulative-incidence-function-cif-equal-to-the-probability-of-observing-an-event-of-interest-by-time-t">1. The cumulative incidence function (CIF) equal to the probability of observing an event of interest by time <span class="math inline">\(t\)</span>
</h4>
<ol type="a">
<li>Use the splitrule <code>logrankCR</code> (the default)</li>
<li>Plot the values for <code>obj$cif</code>
</li>
<li>Use minimal depth for importance</li>
</ol>
<div class="fragment fade-up">
<!-- -->
<h4 id="the-cause-specific-hazard-function-equal-to-the-instantaneous-risk-of-event-of-interest-j-occuring-at-time-t">2. The cause-specific hazard function equal to the instantaneous risk of event of interest <span class="math inline">\(j\)</span> occuring at time <span class="math inline">\(t\)</span>
</h4>
<ol type="a">
<li>Use the splitrule <code>logrank</code> and option <code>cause=j</code>
</li>
<li>Use option <code>importance='permute'</code> to obtain importance and extract column <span class="math inline">\(j\)</span> of <code>obj$importance</code><br>
</li>
</ol>
</div>
</section><section id="competing-risk-example-pbc-mayo-clinic" class="slide level2 smaller"><h2>Competing risk example: PBC Mayo Clinic</h2>
<div class="cell">
<div class="cell-output-display">
<div id="izajgjqssh" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>#izajgjqssh table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#izajgjqssh thead, #izajgjqssh tbody, #izajgjqssh tfoot, #izajgjqssh tr, #izajgjqssh td, #izajgjqssh th {
  border-style: none;
}

#izajgjqssh p {
  margin: 0;
  padding: 0;
}

#izajgjqssh .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#izajgjqssh .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#izajgjqssh .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#izajgjqssh .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#izajgjqssh .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#izajgjqssh .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#izajgjqssh .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#izajgjqssh .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#izajgjqssh .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#izajgjqssh .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#izajgjqssh .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#izajgjqssh .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#izajgjqssh .gt_spanner_row {
  border-bottom-style: hidden;
}

#izajgjqssh .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#izajgjqssh .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#izajgjqssh .gt_from_md > :first-child {
  margin-top: 0;
}

#izajgjqssh .gt_from_md > :last-child {
  margin-bottom: 0;
}

#izajgjqssh .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#izajgjqssh .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#izajgjqssh .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#izajgjqssh .gt_row_group_first td {
  border-top-width: 2px;
}

#izajgjqssh .gt_row_group_first th {
  border-top-width: 2px;
}

#izajgjqssh .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#izajgjqssh .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#izajgjqssh .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#izajgjqssh .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#izajgjqssh .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#izajgjqssh .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#izajgjqssh .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#izajgjqssh .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#izajgjqssh .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#izajgjqssh .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#izajgjqssh .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#izajgjqssh .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#izajgjqssh .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#izajgjqssh .gt_left {
  text-align: left;
}

#izajgjqssh .gt_center {
  text-align: center;
}

#izajgjqssh .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#izajgjqssh .gt_font_normal {
  font-weight: normal;
}

#izajgjqssh .gt_font_bold {
  font-weight: bold;
}

#izajgjqssh .gt_font_italic {
  font-style: italic;
}

#izajgjqssh .gt_super {
  font-size: 65%;
}

#izajgjqssh .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#izajgjqssh .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#izajgjqssh .gt_indent_1 {
  text-indent: 5px;
}

#izajgjqssh .gt_indent_2 {
  text-indent: 10px;
}

#izajgjqssh .gt_indent_3 {
  text-indent: 15px;
}

#izajgjqssh .gt_indent_4 {
  text-indent: 20px;
}

#izajgjqssh .gt_indent_5 {
  text-indent: 25px;
}
</style>
<table class="gt_table caption-top" data-quarto-postprocess="true" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<thead><tr class="header gt_col_headings">
<th id="time" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">time</th>
<th id="status" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">status</th>
<th id="trt" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">trt</th>
<th id="age" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">age</th>
<th id="sex" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col">sex</th>
<th id="ascites" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">ascites</th>
<th id="hepato" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">hepato</th>
<th id="spiders" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">spiders</th>
<th id="edema" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">edema</th>
<th id="bili" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">bili</th>
<th id="chol" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">chol</th>
<th id="albumin" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">albumin</th>
<th id="copper" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">copper</th>
<th id="alk.phos" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">alk.phos</th>
<th id="ast" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">ast</th>
<th id="trig" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">trig</th>
<th id="platelet" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">platelet</th>
<th id="protime" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">protime</th>
</tr></thead>
<tbody class="gt_table_body">
<tr class="odd">
<td class="gt_row gt_right" headers="time">400</td>
<td class="gt_row gt_center" headers="status" style="background-color: #ADD8E6">2</td>
<td class="gt_row gt_right" headers="trt">1</td>
<td class="gt_row gt_right" headers="age">58.76523</td>
<td class="gt_row gt_left" headers="sex">f</td>
<td class="gt_row gt_right" headers="ascites">1</td>
<td class="gt_row gt_right" headers="hepato">1</td>
<td class="gt_row gt_right" headers="spiders">1</td>
<td class="gt_row gt_right" headers="edema">1.0</td>
<td class="gt_row gt_right" headers="bili">14.5</td>
<td class="gt_row gt_right" headers="chol">261</td>
<td class="gt_row gt_right" headers="albumin">2.60</td>
<td class="gt_row gt_right" headers="copper">156</td>
<td class="gt_row gt_right" headers="alk.phos">1718.0</td>
<td class="gt_row gt_right" headers="ast">137.95</td>
<td class="gt_row gt_right" headers="trig">172</td>
<td class="gt_row gt_right" headers="platelet">190</td>
<td class="gt_row gt_right" headers="protime">12.2</td>
</tr>
<tr class="even">
<td class="gt_row gt_right" headers="time">4500</td>
<td class="gt_row gt_center" headers="status" style="background-color: #ADD8E6">0</td>
<td class="gt_row gt_right" headers="trt">1</td>
<td class="gt_row gt_right" headers="age">56.44627</td>
<td class="gt_row gt_left" headers="sex">f</td>
<td class="gt_row gt_right" headers="ascites">0</td>
<td class="gt_row gt_right" headers="hepato">1</td>
<td class="gt_row gt_right" headers="spiders">1</td>
<td class="gt_row gt_right" headers="edema">0.0</td>
<td class="gt_row gt_right" headers="bili">1.1</td>
<td class="gt_row gt_right" headers="chol">302</td>
<td class="gt_row gt_right" headers="albumin">4.14</td>
<td class="gt_row gt_right" headers="copper">54</td>
<td class="gt_row gt_right" headers="alk.phos">7394.8</td>
<td class="gt_row gt_right" headers="ast">113.52</td>
<td class="gt_row gt_right" headers="trig">88</td>
<td class="gt_row gt_right" headers="platelet">221</td>
<td class="gt_row gt_right" headers="protime">10.6</td>
</tr>
<tr class="odd">
<td class="gt_row gt_right" headers="time">1012</td>
<td class="gt_row gt_center" headers="status" style="background-color: #ADD8E6">2</td>
<td class="gt_row gt_right" headers="trt">1</td>
<td class="gt_row gt_right" headers="age">70.07255</td>
<td class="gt_row gt_left" headers="sex">m</td>
<td class="gt_row gt_right" headers="ascites">0</td>
<td class="gt_row gt_right" headers="hepato">0</td>
<td class="gt_row gt_right" headers="spiders">0</td>
<td class="gt_row gt_right" headers="edema">0.5</td>
<td class="gt_row gt_right" headers="bili">1.4</td>
<td class="gt_row gt_right" headers="chol">176</td>
<td class="gt_row gt_right" headers="albumin">3.48</td>
<td class="gt_row gt_right" headers="copper">210</td>
<td class="gt_row gt_right" headers="alk.phos">516.0</td>
<td class="gt_row gt_right" headers="ast">96.10</td>
<td class="gt_row gt_right" headers="trig">55</td>
<td class="gt_row gt_right" headers="platelet">151</td>
<td class="gt_row gt_right" headers="protime">12.0</td>
</tr>
<tr class="even">
<td class="gt_row gt_right" headers="time">1925</td>
<td class="gt_row gt_center" headers="status" style="background-color: #ADD8E6">2</td>
<td class="gt_row gt_right" headers="trt">1</td>
<td class="gt_row gt_right" headers="age">54.74059</td>
<td class="gt_row gt_left" headers="sex">f</td>
<td class="gt_row gt_right" headers="ascites">0</td>
<td class="gt_row gt_right" headers="hepato">1</td>
<td class="gt_row gt_right" headers="spiders">1</td>
<td class="gt_row gt_right" headers="edema">0.5</td>
<td class="gt_row gt_right" headers="bili">1.8</td>
<td class="gt_row gt_right" headers="chol">244</td>
<td class="gt_row gt_right" headers="albumin">2.54</td>
<td class="gt_row gt_right" headers="copper">64</td>
<td class="gt_row gt_right" headers="alk.phos">6121.8</td>
<td class="gt_row gt_right" headers="ast">60.63</td>
<td class="gt_row gt_right" headers="trig">92</td>
<td class="gt_row gt_right" headers="platelet">183</td>
<td class="gt_row gt_right" headers="protime">10.3</td>
</tr>
<tr class="odd">
<td class="gt_row gt_right" headers="time">1504</td>
<td class="gt_row gt_center" headers="status" style="background-color: #ADD8E6">1</td>
<td class="gt_row gt_right" headers="trt">2</td>
<td class="gt_row gt_right" headers="age">38.10541</td>
<td class="gt_row gt_left" headers="sex">f</td>
<td class="gt_row gt_right" headers="ascites">0</td>
<td class="gt_row gt_right" headers="hepato">1</td>
<td class="gt_row gt_right" headers="spiders">1</td>
<td class="gt_row gt_right" headers="edema">0.0</td>
<td class="gt_row gt_right" headers="bili">3.4</td>
<td class="gt_row gt_right" headers="chol">279</td>
<td class="gt_row gt_right" headers="albumin">3.53</td>
<td class="gt_row gt_right" headers="copper">143</td>
<td class="gt_row gt_right" headers="alk.phos">671.0</td>
<td class="gt_row gt_right" headers="ast">113.15</td>
<td class="gt_row gt_right" headers="trig">72</td>
<td class="gt_row gt_right" headers="platelet">136</td>
<td class="gt_row gt_right" headers="protime">10.9</td>
</tr>
<tr class="even">
<td class="gt_row gt_right" headers="time">2503</td>
<td class="gt_row gt_center" headers="status" style="background-color: #ADD8E6">2</td>
<td class="gt_row gt_right" headers="trt">2</td>
<td class="gt_row gt_right" headers="age">66.25873</td>
<td class="gt_row gt_left" headers="sex">f</td>
<td class="gt_row gt_right" headers="ascites">0</td>
<td class="gt_row gt_right" headers="hepato">1</td>
<td class="gt_row gt_right" headers="spiders">0</td>
<td class="gt_row gt_right" headers="edema">0.0</td>
<td class="gt_row gt_right" headers="bili">0.8</td>
<td class="gt_row gt_right" headers="chol">248</td>
<td class="gt_row gt_right" headers="albumin">3.98</td>
<td class="gt_row gt_right" headers="copper">50</td>
<td class="gt_row gt_right" headers="alk.phos">944.0</td>
<td class="gt_row gt_right" headers="ast">93.00</td>
<td class="gt_row gt_right" headers="trig">63</td>
<td class="gt_row gt_right" headers="platelet">NA</td>
<td class="gt_row gt_right" headers="protime">11.0</td>
</tr>
<tr class="odd">
<td class="gt_row gt_right" headers="time">1832</td>
<td class="gt_row gt_center" headers="status" style="background-color: #ADD8E6">0</td>
<td class="gt_row gt_right" headers="trt">2</td>
<td class="gt_row gt_right" headers="age">55.53457</td>
<td class="gt_row gt_left" headers="sex">f</td>
<td class="gt_row gt_right" headers="ascites">0</td>
<td class="gt_row gt_right" headers="hepato">1</td>
<td class="gt_row gt_right" headers="spiders">0</td>
<td class="gt_row gt_right" headers="edema">0.0</td>
<td class="gt_row gt_right" headers="bili">1.0</td>
<td class="gt_row gt_right" headers="chol">322</td>
<td class="gt_row gt_right" headers="albumin">4.09</td>
<td class="gt_row gt_right" headers="copper">52</td>
<td class="gt_row gt_right" headers="alk.phos">824.0</td>
<td class="gt_row gt_right" headers="ast">60.45</td>
<td class="gt_row gt_right" headers="trig">213</td>
<td class="gt_row gt_right" headers="platelet">204</td>
<td class="gt_row gt_right" headers="protime">9.7</td>
</tr>
<tr class="even">
<td class="gt_row gt_right" headers="time">2466</td>
<td class="gt_row gt_center" headers="status" style="background-color: #ADD8E6">2</td>
<td class="gt_row gt_right" headers="trt">2</td>
<td class="gt_row gt_right" headers="age">53.05681</td>
<td class="gt_row gt_left" headers="sex">f</td>
<td class="gt_row gt_right" headers="ascites">0</td>
<td class="gt_row gt_right" headers="hepato">0</td>
<td class="gt_row gt_right" headers="spiders">0</td>
<td class="gt_row gt_right" headers="edema">0.0</td>
<td class="gt_row gt_right" headers="bili">0.3</td>
<td class="gt_row gt_right" headers="chol">280</td>
<td class="gt_row gt_right" headers="albumin">4.00</td>
<td class="gt_row gt_right" headers="copper">52</td>
<td class="gt_row gt_right" headers="alk.phos">4651.2</td>
<td class="gt_row gt_right" headers="ast">28.38</td>
<td class="gt_row gt_right" headers="trig">189</td>
<td class="gt_row gt_right" headers="platelet">373</td>
<td class="gt_row gt_right" headers="protime">11.0</td>
</tr>
<tr class="odd">
<td class="gt_row gt_right" headers="time">2400</td>
<td class="gt_row gt_center" headers="status" style="background-color: #ADD8E6">2</td>
<td class="gt_row gt_right" headers="trt">1</td>
<td class="gt_row gt_right" headers="age">42.50787</td>
<td class="gt_row gt_left" headers="sex">f</td>
<td class="gt_row gt_right" headers="ascites">0</td>
<td class="gt_row gt_right" headers="hepato">0</td>
<td class="gt_row gt_right" headers="spiders">1</td>
<td class="gt_row gt_right" headers="edema">0.0</td>
<td class="gt_row gt_right" headers="bili">3.2</td>
<td class="gt_row gt_right" headers="chol">562</td>
<td class="gt_row gt_right" headers="albumin">3.08</td>
<td class="gt_row gt_right" headers="copper">79</td>
<td class="gt_row gt_right" headers="alk.phos">2276.0</td>
<td class="gt_row gt_right" headers="ast">144.15</td>
<td class="gt_row gt_right" headers="trig">88</td>
<td class="gt_row gt_right" headers="platelet">251</td>
<td class="gt_row gt_right" headers="protime">11.0</td>
</tr>
<tr class="even">
<td class="gt_row gt_right" headers="time">51</td>
<td class="gt_row gt_center" headers="status" style="background-color: #ADD8E6">2</td>
<td class="gt_row gt_right" headers="trt">2</td>
<td class="gt_row gt_right" headers="age">70.55989</td>
<td class="gt_row gt_left" headers="sex">f</td>
<td class="gt_row gt_right" headers="ascites">1</td>
<td class="gt_row gt_right" headers="hepato">0</td>
<td class="gt_row gt_right" headers="spiders">1</td>
<td class="gt_row gt_right" headers="edema">1.0</td>
<td class="gt_row gt_right" headers="bili">12.6</td>
<td class="gt_row gt_right" headers="chol">200</td>
<td class="gt_row gt_right" headers="albumin">2.74</td>
<td class="gt_row gt_right" headers="copper">140</td>
<td class="gt_row gt_right" headers="alk.phos">918.0</td>
<td class="gt_row gt_right" headers="ast">147.25</td>
<td class="gt_row gt_right" headers="trig">143</td>
<td class="gt_row gt_right" headers="platelet">302</td>
<td class="gt_row gt_right" headers="protime">11.5</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div class="sourceCode" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href=""></a><span class="fu">data</span>(pbc, <span class="at">package =</span> <span class="st">"survival"</span>)  </span>
<span id="cb7-2"><a href=""></a></span>
<span id="cb7-3"><a href=""></a>pbc<span class="sc">$</span>id <span class="ot">&lt;-</span> <span class="cn">NULL</span> <span class="do">## remove the ID</span></span>
<span id="cb7-4"><a href=""></a></span>
<span id="cb7-5"><a href=""></a><span class="do">## status at endpoint, 0/1/2 for censored, transplant, dead</span></span>
<span id="cb7-6"><a href=""></a>pbc.cr <span class="ot">&lt;-</span> pbc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section><section id="competing-risk-example-pbc-mayo-clinic-1" class="slide level2 smaller"><h2>Competing risk example: PBC Mayo Clinic</h2>
<div class="sourceCode" id="cb8" data-code-line-numbers="2|13-14|17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href=""></a><span class="do">## canonical example: default Gray's splitting-rule</span></span>
<span id="cb8-2"><a href=""></a>o <span class="ot">&lt;-</span> <span class="fu">rfsrc</span>(<span class="fu">Surv</span>(time, status) <span class="sc">~</span> ., pbc)</span>
<span id="cb8-3"><a href=""></a>o</span>
<span id="cb8-4"><a href=""></a>                         Sample size<span class="sc">:</span> <span class="dv">276</span></span>
<span id="cb8-5"><a href=""></a>                    Number of events<span class="sc">:</span> <span class="dv">18</span>, <span class="dv">111</span></span>
<span id="cb8-6"><a href=""></a>                     Number of trees<span class="sc">:</span> <span class="dv">500</span></span>
<span id="cb8-7"><a href=""></a>           Forest terminal node size<span class="sc">:</span> <span class="dv">15</span></span>
<span id="cb8-8"><a href=""></a>       Average no. of terminal nodes<span class="sc">:</span> <span class="fl">13.118</span></span>
<span id="cb8-9"><a href=""></a>No. of variables tried at each split<span class="sc">:</span> <span class="dv">5</span></span>
<span id="cb8-10"><a href=""></a>              Total no. of variables<span class="sc">:</span> <span class="dv">17</span></span>
<span id="cb8-11"><a href=""></a>       Resampling used to grow trees<span class="sc">:</span> swor</span>
<span id="cb8-12"><a href=""></a>    Resample size used to grow trees<span class="sc">:</span> <span class="dv">174</span></span>
<span id="cb8-13"><a href=""></a>                            Analysis<span class="sc">:</span> RSF</span>
<span id="cb8-14"><a href=""></a>                              Family<span class="sc">:</span> surv<span class="sc">-</span>CR</span>
<span id="cb8-15"><a href=""></a>                      Splitting rule<span class="sc">:</span> logrankCR <span class="sc">*</span>random<span class="sc">*</span></span>
<span id="cb8-16"><a href=""></a>       Number of random split points<span class="sc">:</span> <span class="dv">10</span></span>
<span id="cb8-17"><a href=""></a>   (OOB) Requested performance error<span class="sc">:</span> <span class="fl">0.18408881</span>, <span class="fl">0.1666207</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Tip</strong></p>
</div>
<div class="callout-content">
<p><code>rfsrc</code> recognize competing risk problem automatically when using the survival formula</p>
</div>
</div>
</div>
</section><section id="event-specific-and-non-event-specific-variable-selection" class="slide level2 smaller"><h2>event specific and non-event specific variable selection</h2>
<div class="sourceCode" id="cb9" data-code-line-numbers="1-8|10-12|16-19"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href=""></a><span class="do">## canonical example: default Gray's splitting-rule</span></span>
<span id="cb9-2"><a href=""></a><span class="do">## status: 0/1/2 for censored, transplant, dead </span></span>
<span id="cb9-3"><a href=""></a>o <span class="ot">&lt;-</span> <span class="fu">rfsrc</span>(<span class="fu">Surv</span>(time, status) <span class="sc">~</span> ., pbc)</span>
<span id="cb9-4"><a href=""></a></span>
<span id="cb9-5"><a href=""></a><span class="do">## log-rank splitting where each event type is treated as the event of interest </span></span>
<span id="cb9-6"><a href=""></a><span class="do">## log-rank cause-1 specific splitting and targeted VIMP for cause 1="transplant"</span></span>
<span id="cb9-7"><a href=""></a>o.log1 <span class="ot">&lt;-</span> <span class="fu">rfsrc</span>(<span class="fu">Surv</span>(time, status) <span class="sc">~</span> ., pbc, </span>
<span id="cb9-8"><a href=""></a>                <span class="at">splitrule =</span> <span class="st">"logrank"</span>, <span class="at">cause =</span> <span class="dv">1</span>, <span class="at">importance =</span> <span class="st">"permute"</span>)</span>
<span id="cb9-9"><a href=""></a></span>
<span id="cb9-10"><a href=""></a><span class="do">## log-rank cause-2 specific splitting and targeted VIMP for cause 2="death"</span></span>
<span id="cb9-11"><a href=""></a>o.log2 <span class="ot">&lt;-</span> <span class="fu">rfsrc</span>(<span class="fu">Surv</span>(time, status) <span class="sc">~</span> ., pbc, </span>
<span id="cb9-12"><a href=""></a>                <span class="at">splitrule =</span> <span class="st">"logrank"</span>, <span class="at">cause =</span> <span class="dv">2</span>, <span class="at">importance =</span> <span class="st">"permute"</span>)</span>
<span id="cb9-13"><a href=""></a></span>
<span id="cb9-14"><a href=""></a><span class="do">## extract minimal depth from the Gray split forest: non-event specific</span></span>
<span id="cb9-15"><a href=""></a><span class="do">## extract VIMP from the log-rank forests: event-specific</span></span>
<span id="cb9-16"><a href=""></a>var.perf <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">md =</span> <span class="fu">max.subtree</span>(o)<span class="sc">$</span>order[, <span class="dv">1</span>],</span>
<span id="cb9-17"><a href=""></a>                       <span class="at">vimp1 =</span> <span class="dv">100</span> <span class="sc">*</span> o.log1<span class="sc">$</span>importance[ ,<span class="dv">1</span>],</span>
<span id="cb9-18"><a href=""></a>                       <span class="at">vimp2 =</span> <span class="dv">100</span> <span class="sc">*</span> o.log2<span class="sc">$</span>importance[ ,<span class="dv">2</span>])</span>
<span id="cb9-19"><a href=""></a><span class="fu">print</span>(var.perf[<span class="fu">order</span>(var.perf<span class="sc">$</span>md), ], <span class="at">digits =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section><section id="event-specific-and-non-event-specific-variable-selection-1" class="slide level2 smaller"><h2>event specific and non-event specific variable selection</h2>
<div class="columns">
<div class="column" style="width:60%;">
<div class="sourceCode" id="cb10"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href=""></a><span class="do">## canonical example: default Gray's splitting-rule</span></span>
<span id="cb10-2"><a href=""></a><span class="do">## status: 0/1/2 for censored, transplant, dead </span></span>
<span id="cb10-3"><a href=""></a>o <span class="ot">&lt;-</span> <span class="fu">rfsrc</span>(<span class="fu">Surv</span>(time, status) <span class="sc">~</span> ., pbc)</span>
<span id="cb10-4"><a href=""></a></span>
<span id="cb10-5"><a href=""></a><span class="do">## log-rank splitting where each event type is treated as the event of interest </span></span>
<span id="cb10-6"><a href=""></a><span class="do">## log-rank cause-1 specific splitting and targeted VIMP for cause 1="transplant"</span></span>
<span id="cb10-7"><a href=""></a>o.log1 <span class="ot">&lt;-</span> <span class="fu">rfsrc</span>(<span class="fu">Surv</span>(time, status) <span class="sc">~</span> ., pbc, </span>
<span id="cb10-8"><a href=""></a>                <span class="at">splitrule =</span> <span class="st">"logrank"</span>, <span class="at">cause =</span> <span class="dv">1</span>, <span class="at">importance =</span> <span class="st">"permute"</span>)</span>
<span id="cb10-9"><a href=""></a></span>
<span id="cb10-10"><a href=""></a><span class="do">## log-rank cause-2 specific splitting and targeted VIMP for cause 2="death"</span></span>
<span id="cb10-11"><a href=""></a>o.log2 <span class="ot">&lt;-</span> <span class="fu">rfsrc</span>(<span class="fu">Surv</span>(time, status) <span class="sc">~</span> ., pbc, </span>
<span id="cb10-12"><a href=""></a>                <span class="at">splitrule =</span> <span class="st">"logrank"</span>, <span class="at">cause =</span> <span class="dv">2</span>, <span class="at">importance =</span> <span class="st">"permute"</span>)</span>
<span id="cb10-13"><a href=""></a></span>
<span id="cb10-14"><a href=""></a><span class="do">## extract minimal depth from the Gray split forest: non-event specific</span></span>
<span id="cb10-15"><a href=""></a><span class="do">## extract VIMP from the log-rank forests: event-specific</span></span>
<span id="cb10-16"><a href=""></a>var.perf <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">md =</span> <span class="fu">max.subtree</span>(o)<span class="sc">$</span>order[, <span class="dv">1</span>],</span>
<span id="cb10-17"><a href=""></a>                       <span class="at">vimp1 =</span> <span class="dv">100</span> <span class="sc">*</span> o.log1<span class="sc">$</span>importance[ ,<span class="dv">1</span>],</span>
<span id="cb10-18"><a href=""></a>                       <span class="at">vimp2 =</span> <span class="dv">100</span> <span class="sc">*</span> o.log2<span class="sc">$</span>importance[ ,<span class="dv">2</span>])</span>
<span id="cb10-19"><a href=""></a><span class="fu">print</span>(var.perf[<span class="fu">order</span>(var.perf<span class="sc">$</span>md), ], <span class="at">digits =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div><div class="column fragment fade-left" style="width:40%;">
<div class="sourceCode" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href=""></a></span>
<span id="cb11-2"><a href=""></a>          md  vimp1   vimp2</span>
<span id="cb11-3"><a href=""></a>bili     <span class="fl">2.0</span>  <span class="fl">6.407</span>  <span class="fl">5.9263</span></span>
<span id="cb11-4"><a href=""></a>copper   <span class="fl">3.5</span>  <span class="fl">2.944</span>  <span class="fl">2.0612</span></span>
<span id="cb11-5"><a href=""></a>edema    <span class="fl">4.3</span> <span class="sc">-</span><span class="fl">0.140</span>  <span class="fl">1.1829</span></span>
<span id="cb11-6"><a href=""></a>albumin  <span class="fl">4.3</span> <span class="sc">-</span><span class="fl">0.820</span>  <span class="fl">0.6809</span></span>
<span id="cb11-7"><a href=""></a>ascites  <span class="fl">4.5</span> <span class="sc">-</span><span class="fl">0.012</span>  <span class="fl">1.2883</span></span>
<span id="cb11-8"><a href=""></a>protime  <span class="fl">4.7</span> <span class="sc">-</span><span class="fl">0.832</span>  <span class="fl">0.5708</span></span>
<span id="cb11-9"><a href=""></a>age      <span class="fl">5.0</span>  <span class="fl">7.742</span>  <span class="fl">0.8564</span></span>
<span id="cb11-10"><a href=""></a>chol     <span class="fl">5.0</span>  <span class="fl">0.304</span>  <span class="fl">0.5097</span></span>
<span id="cb11-11"><a href=""></a>ast      <span class="fl">5.6</span> <span class="sc">-</span><span class="fl">0.345</span>  <span class="fl">0.3896</span></span>
<span id="cb11-12"><a href=""></a>alk.phos <span class="fl">6.0</span> <span class="sc">-</span><span class="fl">1.068</span>  <span class="fl">0.0504</span></span>
<span id="cb11-13"><a href=""></a>trig     <span class="fl">6.0</span> <span class="sc">-</span><span class="fl">0.485</span>  <span class="fl">0.0089</span></span>
<span id="cb11-14"><a href=""></a>stage    <span class="fl">6.1</span>  <span class="fl">0.866</span>  <span class="fl">0.3493</span></span>
<span id="cb11-15"><a href=""></a>platelet <span class="fl">6.3</span>  <span class="fl">0.409</span> <span class="sc">-</span><span class="fl">0.1191</span></span>
<span id="cb11-16"><a href=""></a>hepato   <span class="fl">6.6</span>  <span class="fl">0.772</span>  <span class="fl">0.1488</span></span>
<span id="cb11-17"><a href=""></a>spiders  <span class="fl">7.2</span> <span class="sc">-</span><span class="fl">0.121</span>  <span class="fl">0.1349</span></span>
<span id="cb11-18"><a href=""></a>sex      <span class="fl">7.3</span>  <span class="fl">0.125</span>  <span class="fl">0.0289</span></span>
<span id="cb11-19"><a href=""></a>trt      <span class="fl">7.9</span> <span class="sc">-</span><span class="fl">0.217</span> <span class="sc">-</span><span class="fl">0.0174</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div>
</section><section id="cif-stratified-by-edema" class="slide level2 smaller"><h2>CIF stratified by edema</h2>
<div class="sourceCode" id="cb12" data-code-line-numbers="1-3|4-7|8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href=""></a><span class="do">## cumulative incidence function (CIF) for transplant and dead stratified by edema</span></span>
<span id="cb12-2"><a href=""></a>cif <span class="ot">&lt;-</span> o<span class="sc">$</span>cif.oob; Time <span class="ot">&lt;-</span> o<span class="sc">$</span>time.interest</span>
<span id="cb12-3"><a href=""></a>edema <span class="ot">&lt;-</span> o<span class="sc">$</span>xvar<span class="sc">$</span>edema</span>
<span id="cb12-4"><a href=""></a>cif.transplant <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">apply</span>(cif[,,<span class="dv">1</span>][edema <span class="sc">==</span> <span class="dv">0</span>,], <span class="dv">2</span>, mean),</span>
<span id="cb12-5"><a href=""></a>                        <span class="fu">apply</span>(cif[,,<span class="dv">1</span>][edema <span class="sc">&gt;</span> <span class="dv">0</span>,], <span class="dv">2</span>, mean))</span>
<span id="cb12-6"><a href=""></a>cif.dead  <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">apply</span>(cif[,,<span class="dv">2</span>][edema <span class="sc">==</span> <span class="dv">0</span>,], <span class="dv">2</span>, mean),</span>
<span id="cb12-7"><a href=""></a>                   <span class="fu">apply</span>(cif[,,<span class="dv">2</span>][edema <span class="sc">&gt;</span> <span class="dv">0</span>,], <span class="dv">2</span>, mean))</span>
<span id="cb12-8"><a href=""></a><span class="fu">matplot</span>(Time, <span class="fu">cbind</span>(cif.transplant, cif.dead), <span class="at">type =</span> <span class="st">"l"</span>,</span>
<span id="cb12-9"><a href=""></a>        <span class="at">lty =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>), <span class="at">col =</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">2</span>), <span class="at">lwd =</span> <span class="dv">3</span>, <span class="at">ylab =</span> <span class="st">"CIF"</span>)</span>
<span id="cb12-10"><a href=""></a><span class="fu">legend</span>(<span class="st">"topleft"</span>, <span class="at">legend =</span> <span class="fu">c</span>(<span class="st">"Transplant (No Edema)"</span>, <span class="st">"Transplant (Edema)"</span>, </span>
<span id="cb12-11"><a href=""></a>                  <span class="st">"Dead (No Edema)"</span>, <span class="st">"Dead (Edema)"</span>),</span>
<span id="cb12-12"><a href=""></a>       <span class="at">lty =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>), <span class="at">col =</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">2</span>), <span class="at">lwd =</span> <span class="dv">3</span>, <span class="at">cex =</span> <span class="fl">1.1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="fragment fade-left">
<p><img data-src="pic/pbcCRcifStr.png" style="width:40.0%"></p>
</div>
<div class="footer">

</div>
</section><section id="the-run.rfsrc-function-for-an-overview" class="slide level2 smaller"><h2>The run.rfsrc function for an overview</h2>
<div class="sourceCode" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href=""></a><span class="do">## tip!!! perform the same analysis as above with one line</span></span>
<span id="cb13-2"><a href=""></a><span class="fu">run.rfsrc</span>(<span class="fu">Surv</span>(time, status) <span class="sc">~</span> ., pbc, <span class="at">ntree=</span><span class="dv">1000</span>, <span class="at">alpha=</span>.<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="fragment fade-left">
<p><img data-src="pic/pbc_CR_Rrfsrc.png" style="width:80.0%"></p>
</div>
</section><section id="the-run.rfsrc-function-for-an-overview-1" class="slide level2 smaller"><h2>The run.rfsrc function for an overview</h2>

<img data-src="pic/pbc_CR_Rrfsrc.png" style="width:130.0%" class="r-stretch"><div class="footer">

</div>
</section><section id="partial-plot" class="slide level2 smaller"><h2>Partial plot</h2>
<h4 id="using-target-to-specify-the-event-of-interest">using “target” to specify the event of interest</h4>
<div class="sourceCode" id="cb14" data-code-line-numbers="1-3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href=""></a>pbc.cr<span class="sc">$</span>time <span class="ot">&lt;-</span> pbc.cr<span class="sc">$</span>time<span class="sc">/</span><span class="dv">365</span>; o <span class="ot">&lt;-</span> <span class="fu">rfsrc</span>(<span class="fu">Surv</span>(time, status) <span class="sc">~</span> ., pbc.cr)</span>
<span id="cb14-2"><a href=""></a><span class="do">## target = an integer value between 1 and J indicating the event of interest</span></span>
<span id="cb14-3"><a href=""></a><span class="fu">plot.variable</span>(o, <span class="at">target =</span> <span class="dv">1</span>, </span>
<span id="cb14-4"><a href=""></a>              <span class="at">xvar.names =</span> <span class="st">"age"</span>, <span class="at">partial =</span> <span class="cn">TRUE</span>, <span class="at">smooth.lines =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="fragment fade-left">
<p><img data-src="pic/pbcCRpartial1.png" style="width:46.0%"></p>
</div>
</section><section id="partial-plot-1" class="slide level2 smaller"><h2>Partial plot</h2>
<h4 id="using-target-to-specify-the-event-of-interest-1">using “target” to specify the event of interest</h4>
<div class="columns">
<div class="column" style="width:50%;">
<div class="sourceCode" id="cb15" data-code-line-numbers="1-3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href=""></a><span class="do">## target = an integer value between 1 and J indicating the event of interest</span></span>
<span id="cb15-2"><a href=""></a><span class="do">## where J is the number of event types</span></span>
<span id="cb15-3"><a href=""></a><span class="fu">plot.variable</span>(o, <span class="at">target =</span> <span class="dv">1</span>, </span>
<span id="cb15-4"><a href=""></a>              <span class="at">xvar.names =</span> <span class="st">"age"</span>, <span class="at">partial =</span> <span class="cn">TRUE</span>, <span class="at">smooth.lines =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><img data-src="pic/pbcCRpartial1.png" style="width:100.0%"></p>
</div><div class="column fragment fade-left" style="width:50%;">
<div class="sourceCode" id="cb16" data-code-line-numbers="1-3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href=""></a><span class="do">## target = an integer value between 1 and J indicating the event of interest</span></span>
<span id="cb16-2"><a href=""></a><span class="do">## where J is the number of event types</span></span>
<span id="cb16-3"><a href=""></a><span class="fu">plot.variable</span>(o, <span class="at">target =</span> <span class="dv">2</span>, </span>
<span id="cb16-4"><a href=""></a>              <span class="at">xvar.names =</span> <span class="st">"age"</span>, <span class="at">partial =</span> <span class="cn">TRUE</span>, <span class="at">smooth.lines =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="fragment fade-left">
<p><img data-src="pic/pbcCRpartial.png" style="width:100.0%"></p>
</div>
</div></div>
</section><section id="multivariate-analysis" class="slide level2 smaller"><h2>Multivariate analysis</h2>
<p><br></p>
<table class="table-striped border small table-sm caption-top">
<colgroup>
<col style="width: 52%">
<col style="width: 47%">
</colgroup>
<thead><tr class="header">
<th>Family</th>
<th>Example Grow Call with Formula Specification</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><span style="color:#00589b;">Regression <br> Quantile Regression</span></td>
<td>
<code>rfsrc(Ozone~., data = airquality)</code> <br><code>quantreg(mpg~., data = mtcars)</code>
</td>
</tr>
<tr class="even">
<td>
<span style="color:#00589b;">Classification</span> <br><span style="color:#00589b; font-weight: bold;">Imbalanced Two-Class</span>
</td>
<td>
<code>rfsrc(Species~., data = iris)</code> <br><code>imbalanced(status~., data = breast)</code>
</td>
</tr>
<tr class="odd">
<td><span style="color:#00589b;">Survival</span></td>
<td><code>rfsrc(Surv(time, status)~., data = veteran)</code></td>
</tr>
<tr class="even">
<td><span style="color:#00589b;">Competing Risk</span></td>
<td><code>rfsrc(Surv(time, status)~., data = wihs)</code></td>
</tr>
<tr class="odd">
<td><span class="fragment highlight-blue" data-fragment-index="1">Multivariate Regression <br> Multivariate Mixed Regression <br> Multivariate Quantile Regression <br> Multivariate Mixed Quantile Regression</span></td>
<td>
<span class="fragment highlight-blue" data-fragment-index="1">rfsrc(Multivar(mpg, cyl)~., data = mtcars)</span> <br><code>rfsrc(cbind(Species,Sepal.Length)~.,data=iris)</code> <br><code>quantreg(cbind(mpg, cyl)~., data = mtcars)</code> <br><code>quantreg(cbind(Species,Sepal.Length)~.,data=iris)</code>
</td>
</tr>
<tr class="even">
<td>Unsupervised <br> sidClustering <br> Breiman (Shi-Horvath)</td>
<td>
<code>rfsrc(data = mtcars)</code> <br><code>sidClustering(data = mtcars)</code> <br><code>sidClustering(data = mtcars, method = "sh")</code>
</td>
</tr>
</tbody>
</table>
<div class="footer">
<p>Learn more <span class="citation" data-cites="HemantMultiv">[<a href="#/references" role="doc-biblioref" onclick="">3</a>]</span>: <a href="https://www.randomforestsrc.org/articles/mvsplit.html">https://www.randomforestsrc.org/articles/mvsplit.html</a></p>
</div>
</section><section id="multivariate-example-nutrigenomic-study" class="slide level2 smaller"><h2>Multivariate example: Nutrigenomic Study</h2>
<p>Study the effects of five diet treatments on 21 liver lipids and 120 hepatic gene expression in wild-type and PPAR-alpha deficient mice. We regress gene expression, diet and genotype (x-variables) on lipid expressions (multivariate y-responses) <span class="citation" data-cites="martin2007novel">[<a href="#/references" role="doc-biblioref" onclick="">4</a>]</span></p>
<div class="cell">
<div class="cell-output-display">
<div id="gqwbtmstge" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>#gqwbtmstge table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#gqwbtmstge thead, #gqwbtmstge tbody, #gqwbtmstge tfoot, #gqwbtmstge tr, #gqwbtmstge td, #gqwbtmstge th {
  border-style: none;
}

#gqwbtmstge p {
  margin: 0;
  padding: 0;
}

#gqwbtmstge .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#gqwbtmstge .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#gqwbtmstge .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#gqwbtmstge .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#gqwbtmstge .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#gqwbtmstge .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#gqwbtmstge .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#gqwbtmstge .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#gqwbtmstge .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#gqwbtmstge .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#gqwbtmstge .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#gqwbtmstge .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#gqwbtmstge .gt_spanner_row {
  border-bottom-style: hidden;
}

#gqwbtmstge .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#gqwbtmstge .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#gqwbtmstge .gt_from_md > :first-child {
  margin-top: 0;
}

#gqwbtmstge .gt_from_md > :last-child {
  margin-bottom: 0;
}

#gqwbtmstge .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#gqwbtmstge .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#gqwbtmstge .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#gqwbtmstge .gt_row_group_first td {
  border-top-width: 2px;
}

#gqwbtmstge .gt_row_group_first th {
  border-top-width: 2px;
}

#gqwbtmstge .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#gqwbtmstge .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#gqwbtmstge .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#gqwbtmstge .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#gqwbtmstge .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#gqwbtmstge .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#gqwbtmstge .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#gqwbtmstge .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#gqwbtmstge .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#gqwbtmstge .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#gqwbtmstge .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#gqwbtmstge .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#gqwbtmstge .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#gqwbtmstge .gt_left {
  text-align: left;
}

#gqwbtmstge .gt_center {
  text-align: center;
}

#gqwbtmstge .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#gqwbtmstge .gt_font_normal {
  font-weight: normal;
}

#gqwbtmstge .gt_font_bold {
  font-weight: bold;
}

#gqwbtmstge .gt_font_italic {
  font-style: italic;
}

#gqwbtmstge .gt_super {
  font-size: 65%;
}

#gqwbtmstge .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#gqwbtmstge .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#gqwbtmstge .gt_indent_1 {
  text-indent: 5px;
}

#gqwbtmstge .gt_indent_2 {
  text-indent: 10px;
}

#gqwbtmstge .gt_indent_3 {
  text-indent: 15px;
}

#gqwbtmstge .gt_indent_4 {
  text-indent: 20px;
}

#gqwbtmstge .gt_indent_5 {
  text-indent: 25px;
}
</style>
<table class="gt_table caption-top" data-quarto-postprocess="true" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<thead><tr class="header gt_col_headings">
<th id="C14.0" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">C14.0</th>
<th id="C16.0" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">C16.0</th>
<th id="C18.0" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">C18.0</th>
<th id="C16.1n.9" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">C16.1n.9</th>
<th id="C16.1n.7" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">C16.1n.7</th>
<th id="C18.1n.9" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">C18.1n.9</th>
<th id="C18.1n.7" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">C18.1n.7</th>
<th id="diet" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">diet</th>
<th id="genotype" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">genotype</th>
<th id="X36b4" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">X36b4</th>
<th id="ACAT1" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">ACAT1</th>
<th id="ACAT2" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">ACAT2</th>
<th id="ACBP" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">ACBP</th>
<th id="ACC1" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">ACC1</th>
<th id="ACC2" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">ACC2</th>
<th id="ACOTH" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">ACOTH</th>
<th id="ADISP" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">ADISP</th>
<th id="ADSS1" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">ADSS1</th>
<th id="ALDH3" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">ALDH3</th>
</tr></thead>
<tbody class="gt_table_body">
<tr class="odd">
<td class="gt_row gt_right" headers="C14.0">0.34</td>
<td class="gt_row gt_right" headers="C16.0">26.45</td>
<td class="gt_row gt_right" headers="C18.0">10.22</td>
<td class="gt_row gt_right" headers="C16.1n.9">0.35</td>
<td class="gt_row gt_right" headers="C16.1n.7">3.10</td>
<td class="gt_row gt_right" headers="C18.1n.9">16.98</td>
<td class="gt_row gt_right" headers="C18.1n.7">2.41</td>
<td class="gt_row gt_center" headers="diet">lin</td>
<td class="gt_row gt_center" headers="genotype">wt</td>
<td class="gt_row gt_right" headers="X36b4">-0.42</td>
<td class="gt_row gt_right" headers="ACAT1">-0.65</td>
<td class="gt_row gt_right" headers="ACAT2">-0.84</td>
<td class="gt_row gt_right" headers="ACBP">-0.34</td>
<td class="gt_row gt_right" headers="ACC1">-1.29</td>
<td class="gt_row gt_right" headers="ACC2">-1.13</td>
<td class="gt_row gt_right" headers="ACOTH">-0.93</td>
<td class="gt_row gt_right" headers="ADISP">-0.98</td>
<td class="gt_row gt_right" headers="ADSS1">-1.19</td>
<td class="gt_row gt_right" headers="ALDH3">-0.68</td>
</tr>
<tr class="even">
<td class="gt_row gt_right" headers="C14.0">0.38</td>
<td class="gt_row gt_right" headers="C16.0">24.04</td>
<td class="gt_row gt_right" headers="C18.0">9.93</td>
<td class="gt_row gt_right" headers="C16.1n.9">0.55</td>
<td class="gt_row gt_right" headers="C16.1n.7">2.54</td>
<td class="gt_row gt_right" headers="C18.1n.9">20.07</td>
<td class="gt_row gt_right" headers="C18.1n.7">3.92</td>
<td class="gt_row gt_center" headers="diet">sun</td>
<td class="gt_row gt_center" headers="genotype">wt</td>
<td class="gt_row gt_right" headers="X36b4">-0.44</td>
<td class="gt_row gt_right" headers="ACAT1">-0.68</td>
<td class="gt_row gt_right" headers="ACAT2">-0.91</td>
<td class="gt_row gt_right" headers="ACBP">-0.32</td>
<td class="gt_row gt_right" headers="ACC1">-1.23</td>
<td class="gt_row gt_right" headers="ACC2">-1.06</td>
<td class="gt_row gt_right" headers="ACOTH">-0.99</td>
<td class="gt_row gt_right" headers="ADISP">-0.97</td>
<td class="gt_row gt_right" headers="ADSS1">-1.00</td>
<td class="gt_row gt_right" headers="ALDH3">-0.62</td>
</tr>
<tr class="odd">
<td class="gt_row gt_right" headers="C14.0">0.36</td>
<td class="gt_row gt_right" headers="C16.0">23.70</td>
<td class="gt_row gt_right" headers="C18.0">8.96</td>
<td class="gt_row gt_right" headers="C16.1n.9">0.55</td>
<td class="gt_row gt_right" headers="C16.1n.7">2.65</td>
<td class="gt_row gt_right" headers="C18.1n.9">22.89</td>
<td class="gt_row gt_right" headers="C18.1n.7">3.96</td>
<td class="gt_row gt_center" headers="diet">sun</td>
<td class="gt_row gt_center" headers="genotype">wt</td>
<td class="gt_row gt_right" headers="X36b4">-0.48</td>
<td class="gt_row gt_right" headers="ACAT1">-0.74</td>
<td class="gt_row gt_right" headers="ACAT2">-1.10</td>
<td class="gt_row gt_right" headers="ACBP">-0.46</td>
<td class="gt_row gt_right" headers="ACC1">-1.30</td>
<td class="gt_row gt_right" headers="ACC2">-1.09</td>
<td class="gt_row gt_right" headers="ACOTH">-1.06</td>
<td class="gt_row gt_right" headers="ADISP">-1.08</td>
<td class="gt_row gt_right" headers="ADSS1">-1.18</td>
<td class="gt_row gt_right" headers="ALDH3">-0.75</td>
</tr>
<tr class="even">
<td class="gt_row gt_right" headers="C14.0">0.22</td>
<td class="gt_row gt_right" headers="C16.0">25.48</td>
<td class="gt_row gt_right" headers="C18.0">8.14</td>
<td class="gt_row gt_right" headers="C16.1n.9">0.49</td>
<td class="gt_row gt_right" headers="C16.1n.7">2.82</td>
<td class="gt_row gt_right" headers="C18.1n.9">21.92</td>
<td class="gt_row gt_right" headers="C18.1n.7">2.52</td>
<td class="gt_row gt_center" headers="diet">fish</td>
<td class="gt_row gt_center" headers="genotype">wt</td>
<td class="gt_row gt_right" headers="X36b4">-0.45</td>
<td class="gt_row gt_right" headers="ACAT1">-0.69</td>
<td class="gt_row gt_right" headers="ACAT2">-0.65</td>
<td class="gt_row gt_right" headers="ACBP">-0.41</td>
<td class="gt_row gt_right" headers="ACC1">-1.26</td>
<td class="gt_row gt_right" headers="ACC2">-1.09</td>
<td class="gt_row gt_right" headers="ACOTH">-0.93</td>
<td class="gt_row gt_right" headers="ADISP">-1.02</td>
<td class="gt_row gt_right" headers="ADSS1">-1.07</td>
<td class="gt_row gt_right" headers="ALDH3">-0.71</td>
</tr>
<tr class="odd">
<td class="gt_row gt_right" headers="C14.0">0.37</td>
<td class="gt_row gt_right" headers="C16.0">24.80</td>
<td class="gt_row gt_right" headers="C18.0">9.63</td>
<td class="gt_row gt_right" headers="C16.1n.9">0.46</td>
<td class="gt_row gt_right" headers="C16.1n.7">2.85</td>
<td class="gt_row gt_right" headers="C18.1n.9">21.38</td>
<td class="gt_row gt_right" headers="C18.1n.7">2.96</td>
<td class="gt_row gt_center" headers="diet">ref</td>
<td class="gt_row gt_center" headers="genotype">wt</td>
<td class="gt_row gt_right" headers="X36b4">-0.42</td>
<td class="gt_row gt_right" headers="ACAT1">-0.71</td>
<td class="gt_row gt_right" headers="ACAT2">-0.54</td>
<td class="gt_row gt_right" headers="ACBP">-0.38</td>
<td class="gt_row gt_right" headers="ACC1">-1.21</td>
<td class="gt_row gt_right" headers="ACC2">-0.89</td>
<td class="gt_row gt_right" headers="ACOTH">-1.00</td>
<td class="gt_row gt_right" headers="ADISP">-0.95</td>
<td class="gt_row gt_right" headers="ADSS1">-1.08</td>
<td class="gt_row gt_right" headers="ALDH3">-0.76</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div class="columns">
<div class="column" style="width:57%;">
<div class="sourceCode" id="cb17" data-code-line-numbers="1-3|5-6|7-8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href=""></a><span class="fu">data</span>(nutrigenomic, <span class="at">package =</span> <span class="st">"randomForestSRC"</span>)</span>
<span id="cb17-2"><a href=""></a><span class="fu">names</span>(nutrigenomic)</span>
<span id="cb17-3"><a href=""></a><span class="sc">&gt;</span> [<span class="dv">1</span>]  <span class="st">"lipids"</span>   <span class="st">"genes"</span>    <span class="st">"diet"</span>     <span class="st">"genotype"</span></span>
<span id="cb17-4"><a href=""></a></span>
<span id="cb17-5"><a href=""></a><span class="fu">dim</span>(nutrigenomic<span class="sc">$</span>lipids)</span>
<span id="cb17-6"><a href=""></a><span class="sc">&gt;</span> [<span class="dv">1</span>]  <span class="dv">40</span> <span class="dv">21</span></span>
<span id="cb17-7"><a href=""></a><span class="fu">dim</span>(nutrigenomic<span class="sc">$</span>genes)</span>
<span id="cb17-8"><a href=""></a><span class="sc">&gt;</span> [<span class="dv">1</span>]  <span class="dv">40</span> <span class="dv">120</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div><div class="column fragment" style="width:43%;">
<div class="sourceCode" id="cb18" data-code-line-numbers="1-4|6-8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href=""></a><span class="do">## diet and genotype are factors</span></span>
<span id="cb18-2"><a href=""></a><span class="fu">head</span>(nutrigenomic<span class="sc">$</span>diet)</span>
<span id="cb18-3"><a href=""></a><span class="sc">&gt;</span> [<span class="dv">1</span>]  lin  sun  sun  fish ref  coc</span>
<span id="cb18-4"><a href=""></a><span class="sc">&gt;</span> Levels<span class="sc">:</span> coc fish lin ref sun</span>
<span id="cb18-5"><a href=""></a></span>
<span id="cb18-6"><a href=""></a><span class="fu">head</span>(nutrigenomic<span class="sc">$</span>genotype)</span>
<span id="cb18-7"><a href=""></a><span class="sc">&gt;</span> [<span class="dv">1</span>]  wt wt wt wt wt wt</span>
<span id="cb18-8"><a href=""></a><span class="sc">&gt;</span> Levels<span class="sc">:</span> ppar wt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div>
</section><section id="multivariate-example-nutrigenomic-study-1" class="slide level2 smaller"><h2>Multivariate example: Nutrigenomic Study</h2>
<div class="sourceCode" id="cb19" data-code-line-numbers="1-5|7-11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href=""></a><span class="do">## parse into y and x data</span></span>
<span id="cb19-2"><a href=""></a>ydta <span class="ot">&lt;-</span> nutrigenomic<span class="sc">$</span>lipids</span>
<span id="cb19-3"><a href=""></a>xdta <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(nutrigenomic<span class="sc">$</span>genes,</span>
<span id="cb19-4"><a href=""></a>                   <span class="at">diet =</span> nutrigenomic<span class="sc">$</span>diet,</span>
<span id="cb19-5"><a href=""></a>                   <span class="at">genotype =</span> nutrigenomic<span class="sc">$</span>genotype)</span>
<span id="cb19-6"><a href=""></a></span>
<span id="cb19-7"><a href=""></a><span class="do">## multivariate regression forest call</span></span>
<span id="cb19-8"><a href=""></a>o <span class="ot">&lt;-</span> <span class="fu">rfsrc</span>(<span class="fu">get.mv.formula</span>(<span class="fu">colnames</span>(ydta)),</span>
<span id="cb19-9"><a href=""></a>            <span class="fu">data.frame</span>(ydta, xdta),</span>
<span id="cb19-10"><a href=""></a>            <span class="at">importance=</span><span class="cn">TRUE</span>, <span class="at">nsplit =</span> <span class="dv">10</span>,</span>
<span id="cb19-11"><a href=""></a>            <span class="at">splitrule =</span> <span class="st">"mahalanobis"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="fragment fade-up">
<div class="sourceCode" id="cb20" data-code-line-numbers="1,12-13|16-17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href=""></a><span class="sc">&gt;</span> <span class="fu">print</span>(o)</span>
<span id="cb20-2"><a href=""></a>                         Sample size<span class="sc">:</span> <span class="dv">40</span></span>
<span id="cb20-3"><a href=""></a>                     Number of trees<span class="sc">:</span> <span class="dv">500</span></span>
<span id="cb20-4"><a href=""></a>           Forest terminal node size<span class="sc">:</span> <span class="dv">5</span></span>
<span id="cb20-5"><a href=""></a>       Average no. of terminal nodes<span class="sc">:</span> <span class="fl">4.776</span></span>
<span id="cb20-6"><a href=""></a>No. of variables tried at each split<span class="sc">:</span> <span class="dv">41</span></span>
<span id="cb20-7"><a href=""></a>              Total no. of variables<span class="sc">:</span> <span class="dv">122</span></span>
<span id="cb20-8"><a href=""></a>              Total no. of responses<span class="sc">:</span> <span class="dv">21</span></span>
<span id="cb20-9"><a href=""></a>         User has requested response<span class="sc">:</span> C14<span class="fl">.0</span></span>
<span id="cb20-10"><a href=""></a>       Resampling used to grow trees<span class="sc">:</span> swor</span>
<span id="cb20-11"><a href=""></a>    Resample size used to grow trees<span class="sc">:</span> <span class="dv">25</span></span>
<span id="cb20-12"><a href=""></a>                            Analysis<span class="sc">:</span> mRF<span class="sc">-</span>R</span>
<span id="cb20-13"><a href=""></a>                              Family<span class="sc">:</span> regr<span class="sc">+</span></span>
<span id="cb20-14"><a href=""></a>                      Splitting rule<span class="sc">:</span> mahalanobis <span class="sc">*</span>random<span class="sc">*</span></span>
<span id="cb20-15"><a href=""></a>       Number of random split points<span class="sc">:</span> <span class="dv">10</span></span>
<span id="cb20-16"><a href=""></a>                     (OOB) R squared<span class="sc">:</span> <span class="fl">0.13227364</span></span>
<span id="cb20-17"><a href=""></a>   (OOB) Requested performance error<span class="sc">:</span> <span class="fl">0.55613339</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="multivariate-example-nutrigenomic-study-2" class="slide level2 smaller"><h2>Multivariate example: Nutrigenomic Study</h2>
<div class="sourceCode" id="cb21"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href=""></a><span class="do">## acquire the error rate for each of the 21-coordinates </span></span>
<span id="cb21-2"><a href=""></a><span class="do">## standardize to allow for comparison across coordinates</span></span>
<span id="cb21-3"><a href=""></a>serr <span class="ot">&lt;-</span> <span class="fu">get.mv.error</span>(o, <span class="at">standardize =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="fragment fade-up fade-in-then-semi-out">
<div class="sourceCode" id="cb22"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href=""></a><span class="sc">&gt;</span> serr</span>
<span id="cb22-2"><a href=""></a>    C14<span class="fl">.0</span>     C16<span class="fl">.0</span>     C18<span class="fl">.0</span>  C16<span class="fl">.1</span>n<span class="fl">.9</span>  C16<span class="fl">.1</span>n<span class="fl">.7</span>  C18<span class="fl">.1</span>n<span class="fl">.9</span>  C18<span class="fl">.1</span>n<span class="fl">.7</span>  C20<span class="fl">.1</span>n<span class="fl">.9</span>  C20<span class="fl">.3</span>n<span class="fl">.9</span>  C18<span class="fl">.2</span>n<span class="fl">.6</span>  C18<span class="fl">.3</span>n<span class="fl">.6</span> </span>
<span id="cb22-3"><a href=""></a><span class="fl">0.8671651</span> <span class="fl">0.6519146</span> <span class="fl">0.6632005</span> <span class="fl">0.7146180</span> <span class="fl">0.8788133</span> <span class="fl">0.7754928</span> <span class="fl">0.8148566</span> <span class="fl">0.7766450</span> <span class="fl">0.8629866</span> <span class="fl">0.8355279</span> <span class="fl">0.9773675</span> </span>
<span id="cb22-4"><a href=""></a> C20<span class="fl">.2</span>n<span class="fl">.6</span>  C20<span class="fl">.3</span>n<span class="fl">.6</span>  C20<span class="fl">.4</span>n<span class="fl">.6</span>  C22<span class="fl">.4</span>n<span class="fl">.6</span>  C22<span class="fl">.5</span>n<span class="fl">.6</span>  C18<span class="fl">.3</span>n<span class="fl">.3</span>  C20<span class="fl">.3</span>n<span class="fl">.3</span>  C20<span class="fl">.5</span>n<span class="fl">.3</span>  C22<span class="fl">.5</span>n<span class="fl">.3</span>  C22<span class="fl">.6</span>n<span class="fl">.3</span> </span>
<span id="cb22-5"><a href=""></a><span class="fl">0.8439182</span> <span class="fl">0.7456620</span> <span class="fl">0.8378708</span> <span class="fl">0.8915473</span> <span class="fl">0.8949931</span> <span class="fl">0.9301005</span> <span class="fl">0.9438823</span> <span class="fl">0.6832386</span> <span class="fl">0.8470016</span> <span class="fl">0.5416053</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="fragment fade-up">
<div class="sourceCode" id="cb23"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href=""></a><span class="do">## acquire standardized VIMP </span></span>
<span id="cb23-2"><a href=""></a>svimp <span class="ot">&lt;-</span> <span class="fu">get.mv.vimp</span>(obj, <span class="at">standardize =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="fragment fade-up">
<div class="sourceCode" id="cb24"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href=""></a><span class="sc">&gt;</span> <span class="fu">head</span>(svimp)</span>
<span id="cb24-2"><a href=""></a>              C14<span class="fl">.0</span>         C16<span class="fl">.0</span>         C18<span class="fl">.0</span>      C16<span class="fl">.1</span>n<span class="fl">.9</span>     C16<span class="fl">.1</span>n<span class="fl">.7</span>     C18<span class="fl">.1</span>n<span class="fl">.9</span>      C18<span class="fl">.1</span>n<span class="fl">.7</span>      C20<span class="fl">.1</span>n<span class="fl">.9</span>      C20<span class="fl">.3</span>n<span class="fl">.9</span>      C18<span class="fl">.2</span>n<span class="fl">.6</span>      C18<span class="fl">.3</span>n<span class="fl">.6</span></span>
<span id="cb24-3"><a href=""></a>X36b4 <span class="sc">-</span><span class="fl">0.0011838300</span> <span class="sc">-</span><span class="fl">0.0011120623</span> <span class="sc">-</span><span class="fl">2.986438e-03</span> <span class="sc">-</span><span class="fl">0.0023431724</span> <span class="sc">-</span><span class="fl">0.001224286</span> <span class="sc">-</span><span class="fl">0.002843073</span> <span class="sc">-</span><span class="fl">0.0015607315</span> <span class="sc">-</span><span class="fl">0.0005896246</span> <span class="sc">-</span><span class="fl">0.0000214573</span>  <span class="fl">0.0004547212</span> <span class="sc">-</span><span class="fl">0.0006680341</span></span>
<span id="cb24-4"><a href=""></a>ACAT1 <span class="sc">-</span><span class="fl">0.0046534735</span>  <span class="fl">0.0029068448</span>  <span class="fl">9.230050e-04</span>  <span class="fl">0.0022818757</span> <span class="sc">-</span><span class="fl">0.005663879</span>  <span class="fl">0.003297788</span> <span class="sc">-</span><span class="fl">0.0031493826</span> <span class="sc">-</span><span class="fl">0.0016189507</span> <span class="sc">-</span><span class="fl">0.0013414155</span> <span class="sc">-</span><span class="fl">0.0002053584</span> <span class="sc">-</span><span class="fl">0.0010019447</span></span>
<span id="cb24-5"><a href=""></a>ACAT2  <span class="fl">0.0136148438</span>  <span class="fl">0.0081002291</span>  <span class="fl">2.995972e-03</span>  <span class="fl">0.0111641579</span>  <span class="fl">0.008484230</span>  <span class="fl">0.021038781</span>  <span class="fl">0.0173969634</span>  <span class="fl">0.0055616191</span>  <span class="fl">0.0041388219</span>  <span class="fl">0.0072896355</span> <span class="sc">-</span><span class="fl">0.0005133367</span></span>
<span id="cb24-6"><a href=""></a>ACBP   <span class="fl">0.0036475488</span>  <span class="fl">0.0335889422</span>  <span class="fl">1.050479e-02</span>  <span class="fl">0.0046146024</span>  <span class="fl">0.003332181</span>  <span class="fl">0.001875450</span>  <span class="fl">0.0058926949</span>  <span class="fl">0.0127454125</span>  <span class="fl">0.0089004756</span>  <span class="fl">0.0127657144</span> <span class="sc">-</span><span class="fl">0.0004641915</span></span>
<span id="cb24-7"><a href=""></a>ACC1  <span class="sc">-</span><span class="fl">0.0006048965</span> <span class="sc">-</span><span class="fl">0.0002409629</span> <span class="sc">-</span><span class="fl">2.053589e-03</span> <span class="sc">-</span><span class="fl">0.0008960043</span> <span class="sc">-</span><span class="fl">0.001153418</span> <span class="sc">-</span><span class="fl">0.001132043</span> <span class="sc">-</span><span class="fl">0.0006816805</span>  <span class="fl">0.0019148640</span> <span class="sc">-</span><span class="fl">0.0011168252</span>  <span class="fl">0.0005493523</span> <span class="sc">-</span><span class="fl">0.0003399971</span></span>
<span id="cb24-8"><a href=""></a>ACC2   <span class="fl">0.0091886014</span> <span class="sc">-</span><span class="fl">0.0029555916</span>  <span class="fl">2.252395e-05</span>  <span class="fl">0.0002228193</span>  <span class="fl">0.011012313</span>  <span class="fl">0.004644160</span>  <span class="fl">0.0098393504</span> <span class="sc">-</span><span class="fl">0.0004550441</span>  <span class="fl">0.0068958281</span>  <span class="fl">0.0056158617</span>  <span class="fl">0.0015149188</span></span>
<span id="cb24-9"><a href=""></a>           C20<span class="fl">.2</span>n<span class="fl">.6</span>      C20<span class="fl">.3</span>n<span class="fl">.6</span>      C20<span class="fl">.4</span>n<span class="fl">.6</span>      C22<span class="fl">.4</span>n<span class="fl">.6</span>      C22<span class="fl">.5</span>n<span class="fl">.6</span>      C18<span class="fl">.3</span>n<span class="fl">.3</span>      C20<span class="fl">.3</span>n<span class="fl">.3</span>     C20<span class="fl">.5</span>n<span class="fl">.3</span>      C22<span class="fl">.5</span>n<span class="fl">.3</span>     C22<span class="fl">.6</span>n<span class="fl">.3</span></span>
<span id="cb24-10"><a href=""></a>X36b4 <span class="sc">-</span><span class="fl">0.0008007217</span> <span class="sc">-</span><span class="fl">0.0002797032</span> <span class="sc">-</span><span class="fl">9.960998e-05</span> <span class="sc">-</span><span class="fl">0.0016751088</span> <span class="sc">-</span><span class="fl">0.0011467476</span>  <span class="fl">0.0012917637</span>  <span class="fl">0.0009019783</span> <span class="sc">-</span><span class="fl">0.001953023</span> <span class="sc">-</span><span class="fl">0.0028467311</span> <span class="sc">-</span><span class="fl">0.004910317</span></span>
<span id="cb24-11"><a href=""></a>ACAT1  <span class="fl">0.0016278408</span>  <span class="fl">0.0015450041</span>  <span class="fl">1.460450e-03</span>  <span class="fl">0.0039728775</span> <span class="sc">-</span><span class="fl">0.0002436180</span> <span class="sc">-</span><span class="fl">0.0015271289</span> <span class="sc">-</span><span class="fl">0.0009404719</span>  <span class="fl">0.002487745</span>  <span class="fl">0.0085764654</span>  <span class="fl">0.002279551</span></span>
<span id="cb24-12"><a href=""></a>ACAT2  <span class="fl">0.0163623835</span>  <span class="fl">0.0039925515</span>  <span class="fl">1.082762e-02</span>  <span class="fl">0.0243844140</span>  <span class="fl">0.0234540852</span>  <span class="fl">0.0256505212</span>  <span class="fl">0.0340391420</span>  <span class="fl">0.025389222</span>  <span class="fl">0.0117149395</span>  <span class="fl">0.010141345</span></span>
<span id="cb24-13"><a href=""></a>ACBP   <span class="fl">0.0054934515</span>  <span class="fl">0.0086732580</span>  <span class="fl">7.791429e-04</span>  <span class="fl">0.0015539132</span> <span class="sc">-</span><span class="fl">0.0023644673</span>  <span class="fl">0.0211272426</span>  <span class="fl">0.0095212814</span>  <span class="fl">0.005460451</span>  <span class="fl">0.0069928914</span>  <span class="fl">0.009941966</span></span>
<span id="cb24-14"><a href=""></a>ACC1  <span class="sc">-</span><span class="fl">0.0003640483</span> <span class="sc">-</span><span class="fl">0.0024303740</span> <span class="sc">-</span><span class="fl">1.041940e-03</span>  <span class="fl">0.0000992922</span> <span class="sc">-</span><span class="fl">0.0004331203</span> <span class="sc">-</span><span class="fl">0.0008617263</span> <span class="sc">-</span><span class="fl">0.0009875129</span> <span class="sc">-</span><span class="fl">0.001643309</span> <span class="sc">-</span><span class="fl">0.0007186863</span> <span class="sc">-</span><span class="fl">0.003139676</span></span>
<span id="cb24-15"><a href=""></a>ACC2  <span class="sc">-</span><span class="fl">0.0010663573</span> <span class="sc">-</span><span class="fl">0.0007101648</span> <span class="sc">-</span><span class="fl">1.196044e-03</span> <span class="sc">-</span><span class="fl">0.0014820461</span> <span class="sc">-</span><span class="fl">0.0023466329</span>  <span class="fl">0.0017062532</span>  <span class="fl">0.0006796071</span>  <span class="fl">0.002065533</span> <span class="sc">-</span><span class="fl">0.0011808523</span>  <span class="fl">0.003489718</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="split-rules-for-multivariate-regression" class="slide level2 smaller"><h2>Split rules for multivariate regression</h2>
<p><br></p>
<table class="table-striped border small table-sm caption-top">
<colgroup>
<col style="width: 52%">
<col style="width: 47%">
</colgroup>
<thead><tr class="header">
<th>Family</th>
<th><code>splitrule</code></th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Regression <br> Quantile Regression</td>
<td>
<code>mse</code> <br><code>la.quantile.regr, quantile.regr,</code> <code>mse</code>
</td>
</tr>
<tr class="even">
<td>Classification <br> Imbalanced Two-Class</td>
<td>
<code>gini, auc, entropy</code> <br><code>gini, auc, entropy</code>
</td>
</tr>
<tr class="odd">
<td>Survival</td>
<td><code>logrank, bs.gradient, logrankscore</code></td>
</tr>
<tr class="even">
<td>Competing Risk</td>
<td><code>logrankCR, logrank</code></td>
</tr>
<tr class="odd">
<td>
<span style="color:#00589b;">Multivariate Regression</span> <br> Multivariate Classification <br> Multivariate Mixed Regression <br> Multivariate Quantile Regression <br> Multivariate Mixed Quantile Regression</td>
<td>
<span style="color:#00589b;">mv.mse</span>, <span style="color:#00589b;">mahalanobis</span> <br><code>mv.gini</code> <br><code>mv.mix</code> <br><span style="color:#00589b;">mv.mse</span> <br><code>mv.mix</code>
</td>
</tr>
<tr class="even">
<td>Unsupervised <br> sidClustering <br> Breiman (Shi-Horvath)</td>
<td>
<code>unsupv</code> <br><span class="math inline">\(\{\)</span><code>mv.mse, mv.gini, mv.mix</code><span class="math inline">\(\}\)</span>, <code>mahalanobis</code> <br><code>gini, auc, entropy</code>
</td>
</tr>
</tbody>
</table></section><section id="split-rules" class="slide level2 smaller"><h2>Split rules</h2>
<h4 id="mahalanobis-splitting">mahalanobis splitting</h4>
<div class="sourceCode" id="cb25" data-code-line-numbers="4,18|20-21"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href=""></a>o <span class="ot">&lt;-</span> <span class="fu">rfsrc</span>(<span class="fu">get.mv.formula</span>(<span class="fu">colnames</span>(ydta)),</span>
<span id="cb25-2"><a href=""></a>             <span class="fu">data.frame</span>(ydta, xdta),</span>
<span id="cb25-3"><a href=""></a>             <span class="at">importance=</span><span class="cn">TRUE</span>, <span class="at">nsplit =</span> <span class="dv">10</span>,</span>
<span id="cb25-4"><a href=""></a>             <span class="at">splitrule =</span> <span class="st">"mahalanobis"</span>)</span>
<span id="cb25-5"><a href=""></a><span class="sc">&gt;</span> <span class="fu">print</span>(o)</span>
<span id="cb25-6"><a href=""></a>                         Sample size<span class="sc">:</span> <span class="dv">40</span></span>
<span id="cb25-7"><a href=""></a>                     Number of trees<span class="sc">:</span> <span class="dv">500</span></span>
<span id="cb25-8"><a href=""></a>           Forest terminal node size<span class="sc">:</span> <span class="dv">5</span></span>
<span id="cb25-9"><a href=""></a>       Average no. of terminal nodes<span class="sc">:</span> <span class="fl">4.776</span></span>
<span id="cb25-10"><a href=""></a>No. of variables tried at each split<span class="sc">:</span> <span class="dv">41</span></span>
<span id="cb25-11"><a href=""></a>              Total no. of variables<span class="sc">:</span> <span class="dv">122</span></span>
<span id="cb25-12"><a href=""></a>              Total no. of responses<span class="sc">:</span> <span class="dv">21</span></span>
<span id="cb25-13"><a href=""></a>         User has requested response<span class="sc">:</span> C14<span class="fl">.0</span></span>
<span id="cb25-14"><a href=""></a>       Resampling used to grow trees<span class="sc">:</span> swor</span>
<span id="cb25-15"><a href=""></a>    Resample size used to grow trees<span class="sc">:</span> <span class="dv">25</span></span>
<span id="cb25-16"><a href=""></a>                            Analysis<span class="sc">:</span> mRF<span class="sc">-</span>R</span>
<span id="cb25-17"><a href=""></a>                              Family<span class="sc">:</span> regr<span class="sc">+</span></span>
<span id="cb25-18"><a href=""></a>                      Splitting rule<span class="sc">:</span> mahalanobis <span class="sc">*</span>random<span class="sc">*</span></span>
<span id="cb25-19"><a href=""></a>       Number of random split points<span class="sc">:</span> <span class="dv">10</span></span>
<span id="cb25-20"><a href=""></a>                     (OOB) R squared<span class="sc">:</span> <span class="fl">0.13227364</span></span>
<span id="cb25-21"><a href=""></a>   (OOB) Requested performance error<span class="sc">:</span> <span class="fl">0.55613339</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section><section id="split-rules-1" class="slide level2 smaller"><h2>Split rules</h2>
<h4 id="default-composite-independence-splitting">default composite (independence) splitting</h4>
<div class="sourceCode" id="cb26" data-code-line-numbers="4,18|20-21"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href=""></a>o2 <span class="ot">&lt;-</span> <span class="fu">rfsrc</span>(<span class="fu">get.mv.formula</span>(<span class="fu">colnames</span>(ydta)),</span>
<span id="cb26-2"><a href=""></a>              <span class="fu">data.frame</span>(ydta, xdta),</span>
<span id="cb26-3"><a href=""></a>              <span class="at">importance=</span><span class="cn">TRUE</span>, <span class="at">nsplit =</span> <span class="dv">10</span>)</span>
<span id="cb26-4"><a href=""></a>              </span>
<span id="cb26-5"><a href=""></a><span class="sc">&gt;</span> <span class="fu">print</span>(o2)</span>
<span id="cb26-6"><a href=""></a>                         Sample size<span class="sc">:</span> <span class="dv">40</span></span>
<span id="cb26-7"><a href=""></a>                     Number of trees<span class="sc">:</span> <span class="dv">500</span></span>
<span id="cb26-8"><a href=""></a>           Forest terminal node size<span class="sc">:</span> <span class="dv">5</span></span>
<span id="cb26-9"><a href=""></a>       Average no. of terminal nodes<span class="sc">:</span> <span class="fl">4.45</span></span>
<span id="cb26-10"><a href=""></a>No. of variables tried at each split<span class="sc">:</span> <span class="dv">41</span></span>
<span id="cb26-11"><a href=""></a>              Total no. of variables<span class="sc">:</span> <span class="dv">122</span></span>
<span id="cb26-12"><a href=""></a>              Total no. of responses<span class="sc">:</span> <span class="dv">21</span></span>
<span id="cb26-13"><a href=""></a>         User has requested response<span class="sc">:</span> C14<span class="fl">.0</span></span>
<span id="cb26-14"><a href=""></a>       Resampling used to grow trees<span class="sc">:</span> swor</span>
<span id="cb26-15"><a href=""></a>    Resample size used to grow trees<span class="sc">:</span> <span class="dv">25</span></span>
<span id="cb26-16"><a href=""></a>                            Analysis<span class="sc">:</span> mRF<span class="sc">-</span>R</span>
<span id="cb26-17"><a href=""></a>                              Family<span class="sc">:</span> regr<span class="sc">+</span></span>
<span id="cb26-18"><a href=""></a>                      Splitting rule<span class="sc">:</span> mv <span class="sc">*</span>random<span class="sc">*</span></span>
<span id="cb26-19"><a href=""></a>       Number of random split points<span class="sc">:</span> <span class="dv">10</span></span>
<span id="cb26-20"><a href=""></a>                     (OOB) R squared<span class="sc">:</span> <span class="fl">0.38963538</span></span>
<span id="cb26-21"><a href=""></a>   (OOB) Requested performance error<span class="sc">:</span> <span class="fl">0.39118801</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section><section id="split-rules-2" class="slide level2 smaller"><h2>Split rules</h2>
<h4 id="compare-vimp-under-the-two-split-rules">compare VIMP under the two split rules</h4>
<div class="sourceCode" id="cb27"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href=""></a><span class="do">## compare standardized VIMP for top 25 variables</span></span>
<span id="cb27-2"><a href=""></a>imp <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">mahalanobis =</span> <span class="fu">rowMeans</span>(<span class="fu">get.mv.vimp</span>(o,  <span class="at">standardize =</span> <span class="cn">TRUE</span>)),</span>
<span id="cb27-3"><a href=""></a>                  <span class="at">default     =</span> <span class="fu">rowMeans</span>(<span class="fu">get.mv.vimp</span>(o2, <span class="at">standardize =</span> <span class="cn">TRUE</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="fragment fade-up">
<div class="sourceCode" id="cb28"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href=""></a><span class="sc">&gt;</span> <span class="fu">print</span>(<span class="dv">100</span> <span class="sc">*</span> imp[<span class="fu">order</span>(imp[,<span class="st">"mahalanobis"</span>], <span class="at">decreasing =</span> <span class="cn">TRUE</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">25</span>], ])</span>
<span id="cb28-2"><a href=""></a>         mahalanobis      default</span>
<span id="cb28-3"><a href=""></a>diet      <span class="fl">10.1554383</span> <span class="fl">50.157098614</span></span>
<span id="cb28-4"><a href=""></a>CYP3A11    <span class="fl">6.9094201</span>  <span class="fl">7.390394227</span></span>
<span id="cb28-5"><a href=""></a>CYP2c29    <span class="fl">4.3809975</span>  <span class="fl">2.318045460</span></span>
<span id="cb28-6"><a href=""></a>PMDCI      <span class="fl">3.0008368</span>  <span class="fl">4.127672741</span></span>
<span id="cb28-7"><a href=""></a>Ntcp       <span class="fl">2.2830390</span>  <span class="fl">1.219654570</span></span>
<span id="cb28-8"><a href=""></a>genotype   <span class="fl">2.0163456</span>  <span class="fl">7.149894438</span></span>
<span id="cb28-9"><a href=""></a>CYP4A10    <span class="fl">1.2867264</span> <span class="sc">-</span><span class="fl">0.032280665</span></span>
<span id="cb28-10"><a href=""></a>SR.BI      <span class="fl">1.2344900</span>  <span class="fl">1.298163431</span></span>
<span id="cb28-11"><a href=""></a>GSTpi2     <span class="fl">1.1666067</span>  <span class="fl">2.810511592</span></span>
<span id="cb28-12"><a href=""></a>THIOL      <span class="fl">1.0794705</span>  <span class="fl">1.230275963</span></span>
<span id="cb28-13"><a href=""></a>CAR1       <span class="fl">1.0716575</span>  <span class="fl">2.320816813</span></span>
<span id="cb28-14"><a href=""></a>SPI1<span class="fl">.1</span>     <span class="fl">0.9334952</span>  <span class="fl">1.622246274</span></span>
<span id="cb28-15"><a href=""></a>G6Pase     <span class="fl">0.8642624</span>  <span class="fl">0.121325455</span></span>
<span id="cb28-16"><a href=""></a>Lpin       <span class="fl">0.8477423</span>  <span class="fl">3.189846540</span></span>
<span id="cb28-17"><a href=""></a>AOX        <span class="fl">0.8048813</span>  <span class="fl">0.004006545</span></span>
<span id="cb28-18"><a href=""></a>mHMGCoAS   <span class="fl">0.7943659</span>  <span class="fl">0.419456736</span></span>
<span id="cb28-19"><a href=""></a>FAT        <span class="fl">0.7777474</span> <span class="sc">-</span><span class="fl">0.045086923</span></span>
<span id="cb28-20"><a href=""></a>PLTP       <span class="fl">0.5652898</span>  <span class="fl">1.217041153</span></span>
<span id="cb28-21"><a href=""></a>ACOTH      <span class="fl">0.5040769</span>  <span class="fl">1.326771272</span></span>
<span id="cb28-22"><a href=""></a>Tpbeta     <span class="fl">0.5003534</span> <span class="sc">-</span><span class="fl">0.054229018</span></span>
<span id="cb28-23"><a href=""></a>L.FABP     <span class="fl">0.4952051</span> <span class="sc">-</span><span class="fl">0.061771625</span></span>
<span id="cb28-24"><a href=""></a>Tpalpha    <span class="fl">0.4547016</span>  <span class="fl">0.221539122</span></span>
<span id="cb28-25"><a href=""></a>ACBP       <span class="fl">0.4307385</span>  <span class="fl">0.985221687</span></span>
<span id="cb28-26"><a href=""></a>BSEP       <span class="fl">0.4185126</span>  <span class="fl">0.248101073</span></span>
<span id="cb28-27"><a href=""></a>FDFT       <span class="fl">0.4046558</span>  <span class="fl">0.235316552</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="footer">

</div>
</section><section id="the-run.rfsrc-function-for-an-overview-2" class="slide level2 smaller"><h2>The run.rfsrc function for an overview</h2>
<div class="sourceCode" id="cb29"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href=""></a><span class="fu">library</span>(randomForestSRC.run)</span>
<span id="cb29-2"><a href=""></a><span class="fu">run.rfsrc</span>(<span class="fu">get.mv.formula</span>(<span class="fu">colnames</span>(ydta)), <span class="fu">data.frame</span>(ydta, xdta))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="fragment">
<p><img data-src="pic/nutrigenomicRrfsrc.png" style="width:90.0%"></p>
</div>
</section><section id="section" class="slide level2"><h2></h2>
<p><img data-src="pic/nutrigenomicRrfsrc.png" height="690"></p>
<div class="footer">

</div>
</section><section id="section-1" class="slide level2"><h2></h2>
<p><img data-src="pic/nutrigenomicVimp.png" height="690"></p>
<div class="footer">

</div>
</section><section id="missing-data-imputation" class="slide level2 smaller"><h2>Missing data imputation</h2>
<p>Missing data can be imputed at various stages of the analysis</p>
<ol type="1">
<li>During training using <code>na.action="na.impute"</code> in <code>rfsrc</code>
</li>
<li>When predicting data using <code>na.action="na.impute"</code> in <code>predict</code>
</li>
<li>Prior to any analysis using <code>impute</code>
</li>
</ol>
<p>There are two methods used for imputing missing data</p>
<ol type="1">
<li>OTFI (On the fly imputation): simultaneously impute data while growing a tree</li>
<li>missForest: grow a forest using each variable as the <span class="math inline">\(y\)</span> outcome and predicting the missing values</li>
</ol></section><section id="general-call-to-impute" class="slide level2"><h2>General call to impute</h2>
<div class="fragment fade-left">
<p><img data-src="pic/impute.png" style="width:80.0%"></p>
</div>
</section><section id="otfi" class="slide level2"><h2>OTFI</h2>
<ol type="1">
<li>Only non-missing data are used for split-statistics. Daughter node membership assignment for missing data is made by sampling from the non-missing inbag data</li>
<li>Can be used for supervised settings</li>
</ol>
<div class="sourceCode" id="cb30"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href=""></a><span class="fu">data</span>(pbc, <span class="at">package =</span> <span class="st">"randomForestSRC"</span>)</span>
<span id="cb30-2"><a href=""></a>pbc.impute <span class="ot">&lt;-</span> <span class="fu">impute</span>(<span class="fu">Surv</span>(days, status)<span class="sc">~</span>.,pbc)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ol start="3" type="1">
<li>Also applicable to unsupervised settings. In this case multivariate splitting is applied to <code>ytry</code> variables which act as pseudo-responses</li>
</ol>
<div class="sourceCode" id="cb31"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href=""></a><span class="fu">data</span>(pbc, <span class="at">package =</span> <span class="st">"randomForestSRC"</span>)</span>
<span id="cb31-2"><a href=""></a>pbc.impute <span class="ot">&lt;-</span> <span class="fu">impute</span>(<span class="at">data=</span>pbc)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section><section id="missforest-and-mforest" class="slide level2"><h2>missForest and mForest</h2>
<h3 id="multivariate-missforest">(multivariate missForest)</h3>
<p>The imputation problem is turned into a series of regression problem</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href=""></a><span class="fu">data</span>(pbc, <span class="at">package =</span> <span class="st">"randomForestSRC"</span>)</span>
<span id="cb32-2"><a href=""></a>pbc.impute <span class="ot">&lt;-</span> <span class="fu">impute</span>(<span class="at">data=</span>pbc, <span class="at">mf.q=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>For datasets with a large number of variables, instead of running <span class="math inline">\(p\)</span> regressions for each cycle, we can group variables into multivariate regressions</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href=""></a><span class="fu">data</span>(housing, <span class="at">package =</span> <span class="st">"randomForestSRC"</span>)</span>
<span id="cb33-2"><a href=""></a>housing.impute <span class="ot">&lt;-</span> <span class="fu">impute</span>(<span class="at">data=</span>housing, <span class="at">mf.q=</span>.<span class="dv">5</span>)</span>
<span id="cb33-3"><a href=""></a>housing.impute <span class="ot">&lt;-</span> <span class="fu">impute</span>(<span class="at">data=</span>housing, <span class="at">mf.q=</span><span class="dv">40</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section><section id="outline" class="slide level2 smaller" data-background-color="azure"><h2>Outline</h2>
<div class="columns">
<div class="column" style="width:47.5%;">
<h4 id="part-i-training"><a href="https://luminwin.github.io/shortCourse/presentationPartI.html">Part I: Training</a></h4>
<ol type="1">
<li>Quick start</li>
<li>Data structures allowed</li>
<li>Training (grow) with examples <br>(regression, classification, survival)</li>
</ol>
<h4 id="part-ii-inference-and-prediction"><a href="https://luminwin.github.io/shortCourse/presentationPartII.html">Part II: Inference and Prediction</a></h4>
<ol type="1">
<li>Inference (OOB)</li>
<li>Prediction Error</li>
<li>Prediction</li>
<li>Restore</li>
<li>Partial Plots</li>
</ol>
</div><div class="column" style="width:5%;">

</div><div class="column" style="width:47.5%;">
<h4 id="part-iii-variable-selection"><a href="https://luminwin.github.io/shortCourse/presentationPartIII.html">Part III: Variable Selection</a></h4>
<ol type="1">
<li>VIMP</li>
<li>Subsampling (Confidence Intervals)</li>
<li>Minimal Depth</li>
<li>VarPro</li>
</ol>
<h4 id="part-iv-advanced-examples"><a href="https://luminwin.github.io/shortCourse/presentationPartIV.html">Part IV: Advanced Examples</a></h4>
<ol type="1">
<li>Class Imbalanced Data</li>
<li>Competing Risks</li>
<li>Multivariate Forests</li>
</ol>
</div></div>
</section><section id="references" class="slide level2 smaller scrollable"><h2>References</h2>
<div class="quarto-auto-generated-content">
<p><img src="images/rfsrc.svg" class="slide-logo"></p>
<div class="footer footer-default">
<p><a href="https://randomforestsrc.org" class="uri">https://randomforestsrc.org</a></p>
</div>
</div>
<div id="refs" class="references csl-bib-body" role="list">
<div id="ref-HemantRFQv" class="csl-entry" role="listitem">
1. Ishwaran H, O’Brien R, Lu M, Kogalur UB. <span class="nocase">randomForestSRC</span>: Random forests quantile classifier <span>(RFQ)</span> vignette. 2021. <a href="http://randomforestsrc.org/articles/imbalance.html">http://randomforestsrc.org/articles/imbalance.html</a>.
</div>
<div id="ref-HemantCompeting" class="csl-entry" role="listitem">
2. Ishwaran H, Gerds TA, Lau BM, Lu M, Kogalur UB. <span class="nocase">randomForestSRC</span>: Competing risks vignette. 2021. <a href="http://randomforestsrc.org/articles/competing.html">http://randomforestsrc.org/articles/competing.html</a>.
</div>
<div id="ref-HemantMultiv" class="csl-entry" role="listitem">
3. Ishwaran H, Tang F, Lu M, Kogalur UB. <span class="nocase">randomForestSRC</span>: Multivariate splitting rule vignette. 2021. <a href="http://randomforestsrc.org/articles/mvsplit.html">http://randomforestsrc.org/articles/mvsplit.html</a>.
</div>
<div id="ref-martin2007novel" class="csl-entry" role="listitem">
4. Martin PG, Guillou H, Lasserre F, Déjean S, Lan A, Pascussi J-M, et al. Novel aspects of PPAR<span class="math inline">\(\alpha\)</span>-mediated regulation of lipid and xenobiotic metabolism revealed through a nutrigenomic study. Hepatology. 2007;45:767–77.
</div>
</div>
</section>
</div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script><script src="presentationPartIV_files/libs/revealjs/dist/reveal.js"></script><!-- reveal.js plugins --><script src="presentationPartIV_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script><script src="presentationPartIV_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script><script src="presentationPartIV_files/libs/revealjs/plugin/reveal-menu/menu.js"></script><script src="presentationPartIV_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script><script src="presentationPartIV_files/libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script><script src="presentationPartIV_files/libs/revealjs/plugin/quarto-support/support.js"></script><script src="presentationPartIV_files/libs/revealjs/plugin/notes/notes.js"></script><script src="presentationPartIV_files/libs/revealjs/plugin/search/search.js"></script><script src="presentationPartIV_files/libs/revealjs/plugin/zoom/zoom.js"></script><script src="presentationPartIV_files/libs/revealjs/plugin/math/math.js"></script><script>window.define = window.backupDefine; window.backupDefine = undefined;</script><script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': true,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":false},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'fade',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'fade',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script><script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script><script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const onCopySuccess = function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      }
      const getTextToCopy = function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
        text: getTextToCopy
      });
      clipboard.on('success', onCopySuccess);
      if (window.document.getElementById('quarto-embedded-source-code-modal')) {
        // For code content inside modals, clipBoardJS needs to be initialized with a container option
        // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
        const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
          text: getTextToCopy,
          container: window.document.getElementById('quarto-embedded-source-code-modal')
        });
        clipboardModal.on('success', onCopySuccess);
      }
        var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
        var mailtoRegex = new RegExp(/^mailto:/);
          var filterRegex = new RegExp('/' + window.location.host + '/');
        var isInternal = (href) => {
            return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
        }
        // Inspect non-navigation links and adorn them if external
     	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
        for (var i=0; i<links.length; i++) {
          const link = links[i];
          if (!isInternal(link.href)) {
            // undo the damage that might have been done by quarto-nav.js in the case of
            // links that we want to consider external
            if (link.dataset.originalHref !== undefined) {
              link.href = link.dataset.originalHref;
            }
          }
        }
      function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
        const config = {
          allowHTML: true,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start',
        };
        if (contentFn) {
          config.content = contentFn;
        }
        if (onTriggerFn) {
          config.onTrigger = onTriggerFn;
        }
        if (onUntriggerFn) {
          config.onUntrigger = onUntriggerFn;
        }
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note) {
            return note.innerHTML;
          } else {
            return "";
          }
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>


</body></html>